<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2024.01.29 -->
        <title>Details of downloading 1kg reference panel - admix-kit docs</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/tabs.css?v=4c969af8" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">admix-kit docs</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">admix-kit docs</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="prepare-dataset.html">Dataset preparation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="cli/index.html">Command line interface</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Command line interface</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="cli/assoc-test.html">Association testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="cli/genet-cor.html">Genetic correlation estimation</a></li>
<li class="toctree-l2"><a class="reference internal" href="cli/pgs.html">Local ancestry-specific polygenic score</a></li>
<li class="toctree-l2"><a class="reference internal" href="cli/simulate-pheno.html">Phenotype simulation</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="api.html">Python API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Python API</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.assoc.marginal.html">admix.assoc.marginal</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.cli.admix_grm.html">admix.cli.admix_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.cli.admix_grm_merge.html">admix.cli.admix_grm_merge</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.cli.genet_cor.html">admix.cli.genet_cor</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.data.calc_partial_pgs.html">admix.data.calc_partial_pgs</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.Dataset.html">admix.Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.admix_geno.html">admix.simulate.admix_geno</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.quant_pheno.html">admix.simulate.quant_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.binary_pheno.html">admix.simulate.binary_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.quant_pheno_1pop.html">admix.simulate.quant_pheno_1pop</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.quant_pheno_grm.html">admix.simulate.quant_pheno_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.data.allele_per_anc.html">admix.data.allele_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.data.af_per_anc.html">admix.data.af_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.io.read_dataset.html">admix.io.read_dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.io.read_lanc.html">admix.io.read_lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.io.read_rfmix.html">admix.io.read_rfmix</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.tools.plink2.lift_over.html">admix.tools.plink2.lift_over</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.plot.lanc.html">admix.plot.lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.plot.admixture.html">admix.plot.admixture</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.plot.manhattan.html">admix.plot.manhattan</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.plot.qq.html">admix.plot.qq</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Example usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="simulate-admix-genotype.html">Admixed genotype simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/analyze-admix-simu-data.html">Basic statistics and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/assoc.html">Association testing (GWAS)</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/genet-cor.html">Genetic correlation across local ancestry segments</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/polygenic-scoring.html">Polygenic scoring for each local ancestry background</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/dataset.html">admix.Dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/select-admixed.html">Selecting admixed individuals</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="edit-this-page">
  <a class="muted-link" href="https://github.com/kangchenghou/admix-kit/edit/main/docs/download-1kg-detail.md" title="Edit this page">
    <svg aria-hidden="true" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <path d="M4 20h4l10.5 -10.5a1.5 1.5 0 0 0 -4 -4l-10.5 10.5v4" />
      <line x1="13.5" y1="6.5" x2="17.5" y2="10.5" />
    </svg>
    <span class="visually-hidden">Edit this page</span>
  </a>
</div><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="details-of-downloading-1kg-reference-panel">
<h1>Details of downloading 1kg reference panel<a class="headerlink" href="#details-of-downloading-1kg-reference-panel" title="Permalink to this heading">#</a></h1>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This provide more details behind <code class="docutils literal notranslate"><span class="pre">admix</span> <span class="pre">get-1kg-ref</span> <span class="pre">--dir=/path/to/1kg-reference</span></code>. However,
running the following code may not result in the exact same file names as the <code class="docutils literal notranslate"><span class="pre">admix</span></code> command.
We suggest you fully understand all the steps before using them.</p>
</div>
<p>Before we start, create a new folder to store these reference data (e.g. <code class="docutils literal notranslate"><span class="pre">/path/to/1kg_pgen</span></code>) and
<code class="docutils literal notranslate"><span class="pre">cd</span> <span class="pre">/path/to/1kg_pgen</span></code>. This will be the directory where we run the following steps
(each step will correspond to a script file).</p>
<p>If you want to skip the details, just save the following 3 steps as 3 separate script files:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">step1-download.pgen.sh</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">step2-divide-chrom.sh</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">step3-download-map.py</span></code></p></li>
</ol>
<p>And run each of them, you will get the 1,000 Genomes project reference panel ready.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>
### Step 1.1: download 1,000 Genomes reference panel

We download the 1,000 Genomes reference panel from [PLINK2 website](https://www.cog-genomics.org/plink/2.0/resources).
````{tab} hg38
```bash
# step1-download-pgen.sh
# ======================
mkdir -p out/pgen &amp;&amp; cd out/pgen
wget https://www.dropbox.com/s/23xlpscis1p5xud/all_hg38_ns.pgen.zst?dl=1 -O hg38.raw.pgen.zst
wget https://www.dropbox.com/s/hy54ba9yvw665xf/all_hg38_ns_noannot.pvar.zst?dl=1 -O hg38.raw.pvar.zst
wget https://www.dropbox.com/s/3j9zg103fi8cjfs/hg38_corrected.psam?dl=1 -O hg38.raw.psam

plink2 --zst-decompress hg38.raw.pgen.zst &gt;hg38.raw.pgen
plink2 --zst-decompress hg38.raw.pvar.zst &gt;hg38.raw.pvar

# related samples
wget https://www.dropbox.com/s/129gx0gl2v7ndg6/deg2_hg38.king.cutoff.out.id?dl=1 -O hg38.king.cutoff.out.id

# basic QC: bi-allelic SNPs, MAC &gt;= 5, chromosome 1-22
plink2 --pfile hg38.raw \
    --allow-extra-chr \
    --rm-dup exclude-all \
    --max-alleles 2 \
    --mac 5 \
    --snps-only \
    --chr 1-22 \
    --set-all-var-ids @:#:\$r:\$a \
    --make-pgen --out hg38

# clean up
rm hg38.*.zst
rm hg38.raw*
cd ../../
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>
````{tab} hg19
```bash
# step1-download-pgen.sh
# ======================
mkdir -p out/pgen &amp;&amp; cd out/pgen
wget https://www.dropbox.com/s/dps1kvlq338ukz8/all_phase3_ns.pgen.zst?dl=1 -O hg19.raw.pgen.zst
wget https://www.dropbox.com/s/uqk3gfhwsvf7bf3/all_phase3_ns_noannot.pvar.zst?dl=1 -O hg19.raw.pvar.zst
wget https://www.dropbox.com/s/6ppo144ikdzery5/phase3_corrected.psam?dl=1 -O hg19.raw.psam

plink2 --zst-decompress hg19.raw.pgen.zst &gt;hg19.raw.pgen
plink2 --zst-decompress hg19.raw.pvar.zst &gt;hg19.raw.pvar

# related samples
wget https://www.dropbox.com/s/zj8d14vv9mp6x3c/deg2_phase3.king.cutoff.out.id?dl=1 -O hg19.king.cutoff.out.id

# basic QC: bi-allelic SNPs, MAC &gt;= 5, chromosome 1-22
plink2 --pfile hg19.raw \
    --allow-extra-chr \
    --rm-dup exclude-all \
    --max-alleles 2 \
    --mac 5 \
    --snps-only \
    --chr 1-22 \
    --set-all-var-ids @:#:\$r:\$a \
    --make-pgen --out hg19

# clean up
rm hg19.*.zst
rm hg19.raw*
cd ../../
```
</pre></div>
</div>
<section id="step-1-2-divide-by-chromosome-to-vcf-files">
<h2>Step 1.2: divide by chromosome to VCF files<a class="headerlink" href="#step-1-2-divide-by-chromosome-to-vcf-files" title="Permalink to this heading">#</a></h2>
<p>This is to export pgen format to VCF files as input to RFmix.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># step2-divide-chrom.sh</span>
<span class="c1"># =====================</span>

<span class="c1"># replace hg38 with hg19 as needed</span>
mkdir<span class="w"> </span>-p<span class="w"> </span>out/vcf
<span class="k">for</span><span class="w"> </span>chrom<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="o">{</span><span class="m">1</span>..22<span class="o">}</span><span class="p">;</span><span class="w"> </span><span class="k">do</span>
<span class="w">    </span>plink2<span class="w"> </span>--pfile<span class="w"> </span>out/pgen/hg38<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--export<span class="w"> </span>vcf<span class="w"> </span>bgz<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--chr<span class="w"> </span><span class="nv">$chrom</span><span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--out<span class="w"> </span>out/vcf/hg38.chr<span class="nv">$chrom</span>
<span class="w">    </span>tabix<span class="w"> </span>-p<span class="w"> </span>vcf<span class="w"> </span>out/vcf/hg38.chr<span class="nv">$chrom</span>.vcf.gz
<span class="k">done</span>
</pre></div>
</div>
</section>
<section id="step-1-3-download-and-process-sample-map-and-genetic-map">
<h2>Step 1.3: download and process sample map and genetic map<a class="headerlink" href="#step-1-3-download-and-process-sample-map-and-genetic-map" title="Permalink to this heading">#</a></h2>
<p>In this step, we download the information for individual in 1,000 Genomes (population information) and
genetic map information.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># step3-download-map.py</span>
<span class="c1"># =====================</span>
<span class="c1"># save this code into a file and</span>
<span class="c1"># call: python3 step3-download-map.py hg38</span>
<span class="c1"># replace hg38 with hg19 as needed</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="c1"># ROOT_DIR contains:</span>
<span class="c1"># - {ROOT_DIR}/pgen/</span>
<span class="c1"># - {ROOT_DIR}/vcf/</span>
<span class="c1"># this script will create</span>
<span class="c1"># - {ROOT_DIR}/metadata/</span>
<span class="n">ROOT_DIR</span> <span class="o">=</span> <span class="s2">&quot;out/&quot;</span>

<span class="k">def</span> <span class="nf">process_sample_map</span><span class="p">(</span><span class="n">build</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Download and format sample map from 1000 Genomes&quot;&quot;&quot;</span>

    <span class="n">sample_map</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ROOT_DIR</span><span class="si">}</span><span class="s2">/pgen/</span><span class="si">{</span><span class="n">build</span><span class="si">}</span><span class="s2">.psam&quot;</span><span class="p">,</span>
        <span class="n">delim_whitespace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">unrelated_id</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ROOT_DIR</span><span class="si">}</span><span class="s2">/pgen/</span><span class="si">{</span><span class="n">build</span><span class="si">}</span><span class="s2">.king.cutoff.out.id&quot;</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ROOT_DIR</span><span class="si">}</span><span class="s2">/metadata&quot;</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">sample_map</span><span class="p">[[</span><span class="s2">&quot;#IID&quot;</span><span class="p">,</span> <span class="s2">&quot;Population&quot;</span><span class="p">,</span> <span class="s2">&quot;SuperPop&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ROOT_DIR</span><span class="si">}</span><span class="s2">/metadata/</span><span class="si">{</span><span class="n">build</span><span class="si">}</span><span class="s2">.full_sample.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">)</span>
    <span class="c1"># filter unrelated</span>
    <span class="n">unrelated_sample_map</span> <span class="o">=</span> <span class="n">sample_map</span><span class="p">[</span><span class="o">~</span><span class="n">sample_map</span><span class="p">[</span><span class="s2">&quot;#IID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">unrelated_id</span><span class="p">[</span><span class="s2">&quot;#IID&quot;</span><span class="p">])]</span>
    <span class="n">unrelated_sample_map</span><span class="p">[[</span><span class="s2">&quot;#IID&quot;</span><span class="p">,</span> <span class="s2">&quot;Population&quot;</span><span class="p">,</span> <span class="s2">&quot;SuperPop&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ROOT_DIR</span><span class="si">}</span><span class="s2">/metadata/</span><span class="si">{</span><span class="n">build</span><span class="si">}</span><span class="s2">.unrelated_sample.tsv&quot;</span><span class="p">,</span>
        <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Population in unrelated sample map:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">unrelated_sample_map</span><span class="p">[</span><span class="s2">&quot;Population&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">())</span>


<span class="k">def</span> <span class="nf">_process_genetic_map</span><span class="p">(</span><span class="n">root_dir</span><span class="p">,</span> <span class="n">build</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Download and format genetic map</span>
<span class="sd">    1. call bash script to download genetic map to out/metadata/genetic_map/raw</span>
<span class="sd">    2. process the genetic map and save to out/metadata/genetic_map</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">assert</span> <span class="n">build</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">],</span> <span class="s2">&quot;build should be hg38 or hg19&quot;</span>
    <span class="n">raw_map_path</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">get_cache_data</span><span class="p">(</span><span class="s2">&quot;genetic_map&quot;</span><span class="p">,</span> <span class="n">build</span><span class="o">=</span><span class="n">build</span><span class="p">)</span>

    <span class="n">raw_map</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
        <span class="n">raw_map_path</span><span class="p">,</span>
        <span class="n">delim_whitespace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/metadata/genetic_map&quot;</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">):</span>
        <span class="n">chrom_map</span> <span class="o">=</span> <span class="n">raw_map</span><span class="p">[</span><span class="n">raw_map</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">]]</span>
        <span class="n">chrom_map</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/metadata/genetic_map/chr</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">.tsv&quot;</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">build</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Received build:&quot;</span><span class="p">,</span> <span class="n">build</span><span class="p">)</span>
    <span class="n">process_sample_map</span><span class="p">(</span><span class="n">build</span><span class="p">)</span>
    <span class="n">process_genetic_map</span><span class="p">(</span><span class="n">build</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>After these 3 steps, check if your folder contains the following files (you would also expect some other additional intermediate files; replace <code class="docutils literal notranslate"><span class="pre">hg38</span></code> with <code class="docutils literal notranslate"><span class="pre">hg19</span></code> as needed).
Then we are ready to run RFmix.</p>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>.
├── out
│   ├── metadata
│   │   ├── genetic_map
│   │   │   ├── chr1.tsv
│   │   │   ├── chr2.tsv
│   │   │   ├── ...
│   │   │   ├── chr22.tsv
│   │   ├── hg38.full_sample.tsv
│   │   └── hg38.unrelated_sample.tsv
│   └── vcf
│       ├── hg38.chr1.vcf.gz
│       ├── hg38.chr1.vcf.gz.tbi
│       ├── ...
│       ├── hg38.chr22.vcf.gz
│       ├── hg38.chr22.vcf.gz.tbi
├── step1-download-pgen.sh
├── step2-divide-chrom.sh
└── step3-download-map.py
</pre></div>
</div>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021-2024, Kangcheng Hou
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/kangchenghou/admix-kit" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Details of downloading 1kg reference panel</a><ul>
<li><a class="reference internal" href="#step-1-2-divide-by-chromosome-to-vcf-files">Step 1.2: divide by chromosome to VCF files</a></li>
<li><a class="reference internal" href="#step-1-3-download-and-process-sample-map-and-genetic-map">Step 1.3: download and process sample map and genetic map</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=b3ba4146"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/scripts/furo.js?v=32e29ea5"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/tabs.js?v=3ee01567"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>