<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2024.01.29 -->
        <title>Estimating genetic correlation - admix-kit docs</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/tabs.css?v=4c969af8" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">admix-kit docs</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">admix-kit docs</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="prepare-dataset.html">Dataset preparation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="cli/index.html">Command line interface</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Command line interface</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="cli/assoc-test.html">Association testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="cli/genet-cor.html">Genetic correlation estimation</a></li>
<li class="toctree-l2"><a class="reference internal" href="cli/pgs.html">Local ancestry-specific polygenic score</a></li>
<li class="toctree-l2"><a class="reference internal" href="cli/simulate-pheno.html">Phenotype simulation</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="api.html">Python API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Python API</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.assoc.marginal.html">admix.assoc.marginal</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.cli.admix_grm.html">admix.cli.admix_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.cli.admix_grm_merge.html">admix.cli.admix_grm_merge</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.cli.genet_cor.html">admix.cli.genet_cor</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.data.calc_partial_pgs.html">admix.data.calc_partial_pgs</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.Dataset.html">admix.Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.admix_geno.html">admix.simulate.admix_geno</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.quant_pheno.html">admix.simulate.quant_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.binary_pheno.html">admix.simulate.binary_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.quant_pheno_1pop.html">admix.simulate.quant_pheno_1pop</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.quant_pheno_grm.html">admix.simulate.quant_pheno_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.data.allele_per_anc.html">admix.data.allele_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.data.af_per_anc.html">admix.data.af_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.io.read_dataset.html">admix.io.read_dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.io.read_lanc.html">admix.io.read_lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.io.read_rfmix.html">admix.io.read_rfmix</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.tools.plink2.lift_over.html">admix.tools.plink2.lift_over</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.plot.lanc.html">admix.plot.lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.plot.admixture.html">admix.plot.admixture</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.plot.manhattan.html">admix.plot.manhattan</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.plot.qq.html">admix.plot.qq</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Example usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="simulate-admix-genotype.html">Admixed genotype simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/analyze-admix-simu-data.html">Basic statistics and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/assoc.html">Association testing (GWAS)</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/genet-cor.html">Genetic correlation across local ancestry segments</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/polygenic-scoring.html">Polygenic scoring for each local ancestry background</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/dataset.html">admix.Dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/select-admixed.html">Selecting admixed individuals</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="edit-this-page">
  <a class="muted-link" href="https://github.com/kangchenghou/admix-kit/edit/main/docs/genet-cor.md" title="Edit this page">
    <svg aria-hidden="true" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <path d="M4 20h4l10.5 -10.5a1.5 1.5 0 0 0 -4 -4l-10.5 10.5v4" />
      <line x1="13.5" y1="6.5" x2="17.5" y2="10.5" />
    </svg>
    <span class="visually-hidden">Edit this page</span>
  </a>
</div><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section class="tex2jax_ignore mathjax_ignore" id="estimating-genetic-correlation">
<h1>Estimating genetic correlation<a class="headerlink" href="#estimating-genetic-correlation" title="Permalink to this heading">#</a></h1>
<p>We use an example to go through the pipeline.</p>
<section id="step-1-download-data">
<h2>Step 1: Download data<a class="headerlink" href="#step-1-download-data" title="Permalink to this heading">#</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget<span class="w"> </span><span class="s2">&quot;https://www.dropbox.com/s/ub9c6l82ek2yq8q/admix-simu-data.zip?dl=1&quot;</span><span class="w"> </span>-O<span class="w"> </span>admix-simu-data.zip
unzip<span class="w"> </span>admix-simu-data.zip
<span class="nv">pfile</span><span class="o">=</span>admix-simu-data/CEU-YRI<span class="w"> </span><span class="c1"># for an example of 3-way admixture, use pfile=admix-simu-data/CEU-YRI-PEL</span>
<span class="nv">out_dir</span><span class="o">=</span>out/
mkdir<span class="w"> </span>-p<span class="w"> </span><span class="si">${</span><span class="nv">out_dir</span><span class="si">}</span>
</pre></div>
</div>
</section>
<section id="step-2-simulate-phenotype">
<h2>Step 2: Simulate phenotype<a class="headerlink" href="#step-2-simulate-phenotype" title="Permalink to this heading">#</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="k">for</span><span class="w"> </span>cor<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">0</span>.9<span class="w"> </span><span class="m">0</span>.95<span class="w"> </span><span class="m">1</span>.0<span class="p">;</span><span class="w"> </span><span class="k">do</span>
<span class="w">    </span>admix<span class="w"> </span>simulate-admix-pheno<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--pfile<span class="w"> </span><span class="si">${</span><span class="nv">pfile</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--hsq<span class="w"> </span><span class="m">0</span>.25<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--p-causal<span class="w"> </span><span class="m">1</span>.0<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--cor<span class="w"> </span><span class="si">${</span><span class="nv">cor</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--n-sim<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--seed<span class="w"> </span><span class="m">1234</span><span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--out-prefix<span class="w"> </span><span class="si">${</span><span class="nv">out_dir</span><span class="si">}</span>/cor-<span class="si">${</span><span class="nv">cor</span><span class="si">}</span>
<span class="k">done</span>
</pre></div>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="k">for</span> <span class="n">cor</span> <span class="ow">in</span> <span class="p">[</span><span class="mf">0.9</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">]:</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/cor-</span><span class="si">{</span><span class="n">cor</span><span class="si">}</span><span class="s2">.pheno&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
        <span class="n">df_sim</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="sa">f</span><span class="s2">&quot;SIM</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="c1"># add random covariates</span>
        <span class="n">df_sim</span><span class="p">[</span><span class="s2">&quot;COVAR&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="n">df_sim</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">df_sim</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/cor-</span><span class="si">{</span><span class="n">cor</span><span class="si">}</span><span class="s2">.sim</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">.pheno&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="step-3-compute-grm">
<h2>Step 3: Compute GRM<a class="headerlink" href="#step-3-compute-grm" title="Permalink to this heading">#</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkdir<span class="w"> </span>-p<span class="w"> </span><span class="si">${</span><span class="nv">out_dir</span><span class="si">}</span>/admix-grm
admix<span class="w"> </span>append-snp-info<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pfile<span class="w"> </span><span class="si">${</span><span class="nv">pfile</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out<span class="w"> </span><span class="si">${</span><span class="nv">pfile</span><span class="si">}</span>.snp_info

admix<span class="w"> </span>admix-grm<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pfile<span class="w"> </span><span class="si">${</span><span class="nv">pfile</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out-prefix<span class="w"> </span><span class="si">${</span><span class="nv">out_dir</span><span class="si">}</span>/admix-grm/grm
</pre></div>
</div>
<section id="step-4-estimate-genetic-correlation">
<h3>Step 4: Estimate genetic correlation<a class="headerlink" href="#step-4-estimate-genetic-correlation" title="Permalink to this heading">#</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">cor</span><span class="o">=</span><span class="m">0</span>.9
<span class="nv">i</span><span class="o">=</span><span class="m">0</span>
mkdir<span class="w"> </span>-p<span class="w"> </span><span class="si">${</span><span class="nv">out_dir</span><span class="si">}</span>/estimate
<span class="c1"># this step will take a while</span>
admix<span class="w"> </span>genet-cor<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pheno<span class="w"> </span><span class="si">${</span><span class="nv">out_dir</span><span class="si">}</span>/cor-<span class="si">${</span><span class="nv">cor</span><span class="si">}</span>.sim<span class="si">${</span><span class="nv">i</span><span class="si">}</span>.pheno<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--grm-prefix<span class="w"> </span><span class="si">${</span><span class="nv">out_dir</span><span class="si">}</span>/admix-grm/grm<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out-dir<span class="w"> </span><span class="si">${</span><span class="nv">out_dir</span><span class="si">}</span>/estimate/cor-<span class="si">${</span><span class="nv">cor</span><span class="si">}</span>.sim<span class="si">${</span><span class="nv">i</span><span class="si">}</span>

admix<span class="w"> </span>summarize-genet-cor<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--est-dir<span class="w"> </span><span class="si">${</span><span class="nv">out_dir</span><span class="si">}</span>/estimate/cor-<span class="si">${</span><span class="nv">cor</span><span class="si">}</span>.sim<span class="si">${</span><span class="nv">i</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out-prefix<span class="w"> </span><span class="si">${</span><span class="nv">out_dir</span><span class="si">}</span>/estimate/cor-<span class="si">${</span><span class="nv">cor</span><span class="si">}</span>.sim<span class="si">${</span><span class="nv">i</span><span class="si">}</span>

cat<span class="w"> </span>out/estimate/cor-0.9.sim0.summary.json
</pre></div>
</div>
<div class="highlight-json notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
<span class="w">    </span><span class="nt">&quot;n&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">5000</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;rg_mode&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.909</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;rg_hpdi(50%)&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">[</span>
<span class="w">        </span><span class="mf">0.854</span><span class="p">,</span>
<span class="w">        </span><span class="mf">0.952</span>
<span class="w">    </span><span class="p">],</span>
<span class="w">    </span><span class="nt">&quot;rg_hpdi(95%)&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">[</span>
<span class="w">        </span><span class="mf">0.729</span><span class="p">,</span>
<span class="w">        </span><span class="mf">1.0</span>
<span class="w">    </span><span class="p">],</span>
<span class="w">    </span><span class="nt">&quot;rg=1_pval&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.13</span>
<span class="p">}</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Here the wide credible interval is due to the small sample size (N=5,000) used in the analysis.</p>
</div>
<p>To obtain results for simulations from all correlation parameters and simulation replicates (or for all traits in real data analysis), we recommend using computing clusters to parallelize this process. After these results are obtained, one can use <code class="docutils literal notranslate"><span class="pre">admix</span> <span class="pre">meta-analyze-genet-cor</span></code> to meta-analyze these results. For example,</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="k">for</span><span class="w"> </span>cor<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">0</span>.9<span class="w"> </span><span class="m">0</span>.95<span class="w"> </span><span class="m">1</span>.0<span class="p">;</span><span class="w"> </span><span class="k">do</span>
<span class="w">    </span>admix<span class="w"> </span>meta-analyze-genet-cor<span class="w"> </span>--loglkl-files<span class="w"> </span><span class="s2">&quot;out/estimate/cor-</span><span class="si">${</span><span class="nv">cor</span><span class="si">}</span><span class="s2">.sim*.loglkl.txt&quot;</span>
<span class="k">done</span>
</pre></div>
</div>
</section>
</section>
<section id="problem-description">
<h2>Problem description<a class="headerlink" href="#problem-description" title="Permalink to this heading">#</a></h2>
<p>Here, we describe how to estimate <span class="math notranslate nohighlight">\(r_\text{admix}\)</span>, the genetic correlation of causal effects across local ancestry backgrounds.</p>
<p>The phenotype of a given admixed individual <span class="math notranslate nohighlight">\(y\)</span> is modeled as function of allelic effect sizes that are allowed to vary across ancestries as:</p>
<div class="math-wrapper docutils container">
<div class="math notranslate nohighlight">
\[
y = \mathbf{g}_1^\top \boldsymbol{\beta}_1 + \mathbf{g}_2^\top \boldsymbol{\beta}_2 + \epsilon, \qquad (1)
\]</div>
</div>
<p>where <span class="math notranslate nohighlight">\(\mathbf{g}_1, \mathbf{g}_2 \in \mathbb{R}^S\)</span> are the ancestry-specific genotypes for local ancestry 1 and 2, and <span class="math notranslate nohighlight">\(\boldsymbol{\beta}_1, \boldsymbol{\beta}_2 \in \mathbb{R}^S\)</span> are the ancestry-specific causal effects (<span class="math notranslate nohighlight">\(S\)</span> is the number of genome-wide causal SNPs).</p>
<p>We assume the ancestry-specific causal effects follow the distribution of</p>
<div class="math-wrapper docutils container">
<div class="math notranslate nohighlight">
\[\begin{split}
\begin{bmatrix} 
\beta_{1s} \\
\beta_{2s}
\end{bmatrix}
\sim \mathcal{N}\left(
\begin{bmatrix}
0\\
0
\end{bmatrix},\begin{bmatrix}
\sigma_{g}^{2}/M &amp; \rho_g/S\\
\rho_g/S &amp; \sigma_{g}^{2}/S
\end{bmatrix}\right),s=1,\dots,S, \qquad (2)
\end{split}\]</div>
</div>
<p>and the genetic correlation is defined as <span class="math notranslate nohighlight">\(r_\text{admix} = \frac{\rho_g}{\sigma_{g}^{2}}\)</span>.</p>
<section id="estimating-r-text-admix">
<h3>Estimating <span class="math notranslate nohighlight">\(r_\text{admix}\)</span><a class="headerlink" href="#estimating-r-text-admix" title="Permalink to this heading">#</a></h3>
<p>Extending Equation (1) to <span class="math notranslate nohighlight">\(N\)</span> individuals with phenotypes <span class="math notranslate nohighlight">\(\mathbf{y} \in \mathbb{R}^N\)</span> and ancestry-specific genotypes <span class="math notranslate nohighlight">\(\mathbf{G}_1, \mathbf{G}_2 \in \mathbb{R}^{N \times S}\)</span>, we have</p>
<div class="math-wrapper docutils container">
<div class="math notranslate nohighlight">
\[
\mathbf{y} = \mathbf{G}_1 \boldsymbol{\beta}_1 + \mathbf{G}_2 \boldsymbol{\beta}_2 + \boldsymbol{\epsilon}, \qquad (3)
\]</div>
</div>
<p>Integrating out <span class="math notranslate nohighlight">\(\boldsymbol{\beta}_1, \boldsymbol{\beta}_2\)</span> and <span class="math notranslate nohighlight">\(\boldsymbol{\epsilon}\)</span>, we have</p>
<div class="math-wrapper docutils container">
<div class="math notranslate nohighlight">
\[
\mathbf{y} \sim \mathcal{N}\left(
\mathbf{0}, \sigma_g^2 \mathbf{K}_1 + \rho_g \mathbf{K}_2 + \sigma_e^2 \mathbf{I} \right), \qquad (4)
\]</div>
</div>
<p>where we have defined the local ancestry-aware GRM matrices are <span class="math notranslate nohighlight">\(\mathbf{K}_1 = \frac{\mathbf{G}_1 \mathbf{G}_1^\top + \mathbf{G}_2 \mathbf{G}_2^\top }{S}, \mathbf{K}_2 = \frac{\mathbf{G}_1 \mathbf{G}_2^\top + \mathbf{G}_2 \mathbf{G}_1^\top }{S}\)</span>. We also note that Equation (4) can be rewritten as</p>
<div class="math-wrapper docutils container">
<div class="math notranslate nohighlight">
\[
\mathbf{y} \sim \mathcal{N}\left(
\mathbf{0}, \sigma_g^2 (\mathbf{K}_1 + r_\text{admix} \mathbf{K}_2) + \sigma_e^2 \mathbf{I} \right). \qquad (4)
\]</div>
</div>
<p>Therefore, we can calculate <span class="math notranslate nohighlight">\(\mathbf{K}(r_\text{admix}) = \mathbf{K}_1 + r_\text{admix} \mathbf{K}_2\)</span> for a grid of values of <span class="math notranslate nohighlight">\(r_\text{admix}\)</span>, and use the likelihood ratio curve to estimate <span class="math notranslate nohighlight">\(r_\text{admix}\)</span>.</p>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021-2024, Kangcheng Hou
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/kangchenghou/admix-kit" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Estimating genetic correlation</a><ul>
<li><a class="reference internal" href="#step-1-download-data">Step 1: Download data</a></li>
<li><a class="reference internal" href="#step-2-simulate-phenotype">Step 2: Simulate phenotype</a></li>
<li><a class="reference internal" href="#step-3-compute-grm">Step 3: Compute GRM</a><ul>
<li><a class="reference internal" href="#step-4-estimate-genetic-correlation">Step 4: Estimate genetic correlation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#problem-description">Problem description</a><ul>
<li><a class="reference internal" href="#estimating-r-text-admix">Estimating <span class="math notranslate nohighlight">\(r_\text{admix}\)</span></a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=b3ba4146"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/scripts/furo.js?v=32e29ea5"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/tabs.js?v=3ee01567"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>