<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Command line interface" href="cli/index.html" /><link rel="prev" title="Overview" href="overview.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2024.01.29 -->
        <title>Dataset preparation - admix-kit docs</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/tabs.css?v=4c969af8" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">admix-kit docs</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">admix-kit docs</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Documentation</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Dataset preparation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="cli/index.html">Command line interface</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Command line interface</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="cli/assoc-test.html">Association testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="cli/genet-cor.html">Genetic correlation estimation</a></li>
<li class="toctree-l2"><a class="reference internal" href="cli/pgs.html">Local ancestry-specific polygenic score</a></li>
<li class="toctree-l2"><a class="reference internal" href="cli/simulate-pheno.html">Phenotype simulation</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="api.html">Python API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Python API</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.assoc.marginal.html">admix.assoc.marginal</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.cli.admix_grm.html">admix.cli.admix_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.cli.admix_grm_merge.html">admix.cli.admix_grm_merge</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.cli.genet_cor.html">admix.cli.genet_cor</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.data.calc_partial_pgs.html">admix.data.calc_partial_pgs</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.Dataset.html">admix.Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.admix_geno.html">admix.simulate.admix_geno</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.quant_pheno.html">admix.simulate.quant_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.binary_pheno.html">admix.simulate.binary_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.quant_pheno_1pop.html">admix.simulate.quant_pheno_1pop</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.simulate.quant_pheno_grm.html">admix.simulate.quant_pheno_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.data.allele_per_anc.html">admix.data.allele_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.data.af_per_anc.html">admix.data.af_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.io.read_dataset.html">admix.io.read_dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.io.read_lanc.html">admix.io.read_lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.io.read_rfmix.html">admix.io.read_rfmix</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.tools.plink2.lift_over.html">admix.tools.plink2.lift_over</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.plot.lanc.html">admix.plot.lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.plot.admixture.html">admix.plot.admixture</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.plot.manhattan.html">admix.plot.manhattan</a></li>
<li class="toctree-l2"><a class="reference internal" href="_api/admix.plot.qq.html">admix.plot.qq</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Example usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="simulate-admix-genotype.html">Admixed genotype simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/analyze-admix-simu-data.html">Basic statistics and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/assoc.html">Association testing (GWAS)</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/genet-cor.html">Genetic correlation across local ancestry segments</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/polygenic-scoring.html">Polygenic scoring for each local ancestry background</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/dataset.html">admix.Dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/select-admixed.html">Selecting admixed individuals</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="edit-this-page">
  <a class="muted-link" href="https://github.com/kangchenghou/admix-kit/edit/main/docs/prepare-dataset.md" title="Edit this page">
    <svg aria-hidden="true" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <path d="M4 20h4l10.5 -10.5a1.5 1.5 0 0 0 -4 -4l-10.5 10.5v4" />
      <line x1="13.5" y1="6.5" x2="17.5" y2="10.5" />
    </svg>
    <span class="visually-hidden">Edit this page</span>
  </a>
</div><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section class="tex2jax_ignore mathjax_ignore" id="dataset-preparation">
<h1>Dataset preparation<a class="headerlink" href="#dataset-preparation" title="Permalink to this heading">#</a></h1>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this heading">#</a></h2>
<p>To start genetic analysis for admixed populations with admix-kit, it is required to compile the data set into the following files.</p>
<ul class="simple">
<li><p>Phased genotype in PLINK2 format.</p></li>
<li><p>Local ancestry inference results .lanc.</p></li>
<li><p>Additional individuals’ information .indiv_info</p></li>
<li><p>Additional SNPs’ information .snp_info</p></li>
</ul>
<p>For genome-wide analysis, it is recommended to divide the data set by chromosomes. For example, the file structure will look like</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>.
├── genotype
│   ├── chr1.pgen
│   ├── chr1.psam
│   ├── chr1.pvar
│   ├── chr1.lanc
│   ├── chr1.indiv_info
│   ├── chr1.snp_info
│   ├── chr2.pgen
│   ├── ...
</pre></div>
</div>
<p>We details the steps to prepare data set as follows:</p>
</section>
<section id="step-1-format-genotype">
<h2>Step 1: format genotype<a class="headerlink" href="#step-1-format-genotype" title="Permalink to this heading">#</a></h2>
<section id="step-1-1-optional-select-well-imputed-snps">
<h3>Step 1.1 (optional): select well-imputed SNPs<a class="headerlink" href="#step-1-1-optional-select-well-imputed-snps" title="Permalink to this heading">#</a></h3>
<p>Often we start with the imputed genotype from imputation server. We can filter by MAF &gt; 0.005 (5th column) and R2 &gt; 0.8 (7th column) to select the SNPs with high quality.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">IN_DIR</span><span class="o">=</span>/path/to/vcf
<span class="nv">OUT_DIR</span><span class="o">=</span>/path/to/imputed

<span class="c1"># convert to PLINK2 format</span>
plink2<span class="w"> </span>--vcf<span class="w"> </span><span class="si">${</span><span class="nv">IN_DIR</span><span class="si">}</span>/chr<span class="si">${</span><span class="nv">chrom</span><span class="si">}</span>.vcf.gz<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--extract-if-info<span class="w"> </span><span class="s2">&quot;R2&gt;0.8&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--rm-dup<span class="w"> </span>exclude-all<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--snps-only<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--maf<span class="w"> </span><span class="m">0</span>.005<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--max-alleles<span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--make-pgen<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--memory<span class="w"> </span><span class="m">16000</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out<span class="w"> </span><span class="si">${</span><span class="nv">OUT_DIR</span><span class="si">}</span>/chr<span class="si">${</span><span class="nv">chrom</span><span class="si">}</span>

<span class="c1"># NOTE: -extract-if-info &quot;R2&gt;0.8&quot; is to retain well-imputed SNPs</span>
<span class="c1"># alternatively, if you R2 information in a .info.gz file</span>
<span class="c1"># you can replace the &#39;--extract-if-info&#39; with</span>
<span class="c1"># $ zcat ${IN_DIR}/chr${chrom}.info.gz | awk &#39;NR&gt;1 {if($5&gt;0.005 &amp;&amp; $7&gt;0.8) print $1}&#39; &gt; ${OUT_DIR}/chr${chrom}.snplist</span>
<span class="c1"># and use plink2 --extract ${OUT_DIR}/chr${chrom}.snplist</span>

<span class="c1"># if your vcf file is already processed, use the following</span>
<span class="c1"># $ plink2 --vcf ${vcf} --make-pgen --out ${out_plink}</span>
</pre></div>
</div>
<p>PLINK2 is also versatile for converting other formats into .pgen format.
See more at <a class="reference external" href="https://www.cog-genomics.org/plink/2.0/input#pgen">https://www.cog-genomics.org/plink/2.0/input#pgen</a>.</p>
</section>
<section id="step-1-2-optional-select-hm3-snps">
<h3>Step 1.2 (optional): select HM3 SNPs<a class="headerlink" href="#step-1-2-optional-select-hm3-snps" title="Permalink to this heading">#</a></h3>
<p>Most genetic analysis (e.g., local ancestry inference) can be made more efficient by subsetting the data to HapMap3 SNPs.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>admix<span class="w"> </span>subset-hapmap3<span class="w"> </span>--pfile<span class="w"> </span><span class="si">${</span><span class="nv">imputed_pfile</span><span class="si">}</span><span class="w"> </span>--out-pfile<span class="w"> </span><span class="si">${</span><span class="nv">hm3_pfile</span><span class="si">}</span><span class="w"> </span>--build<span class="w"> </span>hg38
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Make sure your source data is phased because it is essential for many analyses with admix-kit. Use <code class="docutils literal notranslate"><span class="pre">plink2</span> <span class="pre">--pfile</span> <span class="pre">&lt;pfile&gt;</span> <span class="pre">--pgen-info</span></code> for basic check. If there is a line “Explicitly phased hardcalls present”, that means phasing data is present.</p>
</div>
</section>
<section id="step-1-3-optional-merge-all-chromosomes-into-one-file">
<h3>Step 1.3 (optional): merge all chromosomes into one file<a class="headerlink" href="#step-1-3-optional-merge-all-chromosomes-into-one-file" title="Permalink to this heading">#</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">(</span><span class="k">for</span><span class="w"> </span>i<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="o">{</span><span class="m">1</span>..22<span class="o">}</span><span class="p">;</span><span class="w"> </span><span class="k">do</span><span class="w"> </span><span class="nb">echo</span><span class="w"> </span>$<span class="s2">&quot;chr</span><span class="si">${</span><span class="nv">i</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">;</span><span class="w"> </span><span class="k">done</span><span class="o">)</span><span class="w"> </span>&gt;<span class="w"> </span>chr_list.txt
plink2<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pmerge-list<span class="w"> </span>chr_list.txt<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pmerge-list-dir<span class="w"> </span><span class="si">${</span><span class="nv">pfile_dir</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--make-pgen<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out<span class="w"> </span><span class="si">${</span><span class="nv">merged_pfile</span><span class="si">}</span>
</pre></div>
</div>
</section>
<section id="step-1-4-optional-perform-joint-pca-with-1kg-reference-panel">
<h3>Step 1.4 (optional): perform joint PCA with 1kg reference panel<a class="headerlink" href="#step-1-4-optional-perform-joint-pca-with-1kg-reference-panel" title="Permalink to this heading">#</a></h3>
<p>By overlapping your sample with the 1,000 Genomes reference panel, you can get an overall
idea of genetic ancestries of individuals in your sample.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>admix<span class="w"> </span>pfile-merge-indiv<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pfile1<span class="w"> </span><span class="si">${</span><span class="nv">REF_DIR</span><span class="si">}</span>/all_chr<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pfile2<span class="w"> </span><span class="si">${</span><span class="nv">SAMPLE_HM3_DIR</span><span class="si">}</span>/all_chr<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out<span class="w"> </span><span class="si">${</span><span class="nv">OUT_DIR</span><span class="si">}</span>/merged

plink2<span class="w"> </span>--pfile<span class="w"> </span><span class="si">${</span><span class="nv">OUT_DIR</span><span class="si">}</span>/merged<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pca<span class="w"> </span>approx<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out<span class="w"> </span><span class="si">${</span><span class="nv">OUT_DIR</span><span class="si">}</span>/merged_pca
</pre></div>
</div>
</section>
</section>
<section id="step-2-local-ancestry-inference">
<h2>Step 2: Local ancestry inference<a class="headerlink" href="#step-2-local-ancestry-inference" title="Permalink to this heading">#</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>There are many tools for local ancestry inference.
If you have not performed the local ancestry inference, we prepare a guideline to use RFmix for local ancestry inference (<a class="reference internal" href="rfmix.html"><span class="std std-doc">see RFmix guideline</span></a>).
Otherwise, you can use the tool you like.</p>
</div>
<p>We provide helper function to convert the local ancestry results into .lanc format (see more details below) which is a compact format for storing local ancestry. To convert the RFmix local ancestry into .lanc format, use the following command. This command can be applied to both imputed and hm3 data.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>admix<span class="w"> </span>lanc-convert<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pfile<span class="w"> </span>&lt;pgen_prefix&gt;<span class="w"> </span><span class="se">\ </span><span class="w">     </span><span class="c1"># e.g., dset.chr1</span>
<span class="w">    </span>--rfmix<span class="w"> </span>&lt;rfmix_msp_path&gt;<span class="w"> </span><span class="se">\ </span><span class="w"> </span><span class="c1"># e.g., rfmix/dset.chr1.msp.tsv</span>
<span class="w">    </span>--out<span class="w"> </span>&lt;lanc_path&gt;<span class="w">           </span><span class="c1"># e.g., dset.chr1.lanc</span>
</pre></div>
</div>
<p>Now you already formatted all the required for other downstream analysis. Besides that,
we also recommend calculating some basic statistics for the data set with:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>admix<span class="w"> </span>append-snp-info<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--pfile<span class="w"> </span>&lt;pgen_prefix&gt;<span class="w"> </span><span class="se">\ </span><span class="w">     </span><span class="c1"># e.g., dset.chr1</span>
<span class="w">    </span>--out<span class="w"> </span>&lt;snp_info&gt;<span class="w">  </span><span class="c1"># e.g., dset.chr1.snp_info</span>
</pre></div>
</div>
</section>
<section id="other-files">
<h2>Other files<a class="headerlink" href="#other-files" title="Permalink to this heading">#</a></h2>
<p>PLINK2 genotype file and <code class="docutils literal notranslate"><span class="pre">.lanc</span></code> file are almost you need to start the analysis. The
other files might be useful for better structuring your analysis. <code class="docutils literal notranslate"><span class="pre">.snp_info</span></code> contains
SNP information file, such as allele frequency, and <code class="docutils literal notranslate"><span class="pre">.indiv_info</span></code> contains individual
information file, such as top PCs.</p>
</section>
<section id="example-simulated-dataset">
<h2>Example simulated dataset<a class="headerlink" href="#example-simulated-dataset" title="Permalink to this heading">#</a></h2>
<p>Download an example dataset from <a class="reference internal" href="#"><span class="xref myst">here</span></a>. See <a class="reference internal" href="#"><span class="xref myst">step-by-step instructions</span></a> for simulation.</p>
</section>
<section id="lanc-file-formats">
<h2><code class="docutils literal notranslate"><span class="pre">.lanc</span></code> file formats<a class="headerlink" href="#lanc-file-formats" title="Permalink to this heading">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">.lanc</span></code> is a text file containing a matrix of local ancestry of shape <code class="docutils literal notranslate"><span class="pre">&lt;n_snp&gt;</span> <span class="pre">x</span> <span class="pre">&lt;n_indiv&gt;</span> <span class="pre">x</span> <span class="pre">&lt;2</span> <span class="pre">ploidy&gt;</span></code>.</p>
<p>The first line contains two numbers: <code class="docutils literal notranslate"><span class="pre">&lt;n_snp&gt;</span></code> for number of SNPs and <code class="docutils literal notranslate"><span class="pre">&lt;n_indiv&gt;</span></code> for number of indivduals. Then <code class="docutils literal notranslate"><span class="pre">&lt;n_indiv&gt;</span></code> lines follow with each line corresponds to one individual:
For each line, the local ancestry change points are recorded as
<code class="docutils literal notranslate"><span class="pre">&lt;pos&gt;:&lt;anc1&gt;&lt;anc2&gt;</span></code> which records the position of the change point and the <em>ordered</em> ancestries (according to the phase) local ancestry information.</p>
<p>An example of <code class="docutils literal notranslate"><span class="pre">.lanc</span></code> file will make the format clear:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">300</span> <span class="mi">3</span>
<span class="mi">100</span><span class="p">:</span><span class="mi">01</span> <span class="mi">300</span><span class="p">:</span><span class="mi">00</span>
<span class="mi">120</span><span class="p">:</span><span class="mi">10</span> <span class="mi">300</span><span class="p">:</span><span class="mi">01</span>
<span class="mi">300</span><span class="p">:</span><span class="mi">00</span>
</pre></div>
</div>
<p>This corresponds to a 300 SNPs x 3 individuals x 2 ploidy matrix. The corresponding dense matrix for the first individual can be reconstructed using the following code:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># example for the first individual in the above example file</span>
<span class="n">break_list</span> <span class="o">=</span> <span class="p">[</span><span class="mi">100</span><span class="p">,</span> <span class="mi">300</span><span class="p">]</span>
<span class="n">anc0_list</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="n">anc1_list</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>    
<span class="n">start</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">lanc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)</span>
<span class="k">for</span> <span class="n">stop</span><span class="p">,</span> <span class="n">anc0</span><span class="p">,</span> <span class="n">anc1</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">break_list</span><span class="p">,</span> <span class="n">anc0_list</span><span class="p">,</span> <span class="n">anc1_list</span><span class="p">):</span>
    <span class="n">lanc</span><span class="p">[</span><span class="n">start</span> <span class="p">:</span> <span class="n">stop</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">anc0</span>
    <span class="n">lanc</span><span class="p">[</span><span class="n">start</span> <span class="p">:</span> <span class="n">stop</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">anc1</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">stop</span>
</pre></div>
</div>
<p>Note these ranges are right-open intervals <code class="docutils literal notranslate"><span class="pre">[start,</span> <span class="pre">stop)</span></code> and the last position of each line always ends with <code class="docutils literal notranslate"><span class="pre">&lt;n_snp&gt;</span></code>. We provide helper function to convert between sparse <code class="docutils literal notranslate"><span class="pre">.lanc</span></code> format and dense matrix format.</p>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="cli/index.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Command line interface</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="overview.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Overview</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021-2024, Kangcheng Hou
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/kangchenghou/admix-kit" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Dataset preparation</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#step-1-format-genotype">Step 1: format genotype</a><ul>
<li><a class="reference internal" href="#step-1-1-optional-select-well-imputed-snps">Step 1.1 (optional): select well-imputed SNPs</a></li>
<li><a class="reference internal" href="#step-1-2-optional-select-hm3-snps">Step 1.2 (optional): select HM3 SNPs</a></li>
<li><a class="reference internal" href="#step-1-3-optional-merge-all-chromosomes-into-one-file">Step 1.3 (optional): merge all chromosomes into one file</a></li>
<li><a class="reference internal" href="#step-1-4-optional-perform-joint-pca-with-1kg-reference-panel">Step 1.4 (optional): perform joint PCA with 1kg reference panel</a></li>
</ul>
</li>
<li><a class="reference internal" href="#step-2-local-ancestry-inference">Step 2: Local ancestry inference</a></li>
<li><a class="reference internal" href="#other-files">Other files</a></li>
<li><a class="reference internal" href="#example-simulated-dataset">Example simulated dataset</a></li>
<li><a class="reference internal" href="#lanc-file-formats"><code class="docutils literal notranslate"><span class="pre">.lanc</span></code> file formats</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=b3ba4146"></script>
    <script src="_static/doctools.js?v=888ff710"></script>
    <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/scripts/furo.js?v=32e29ea5"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/tabs.js?v=3ee01567"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>