<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2024.01.29 -->
        <title>admix.data._geno - admix-kit docs</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/tabs.css?v=4c969af8" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">admix-kit docs</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  
  
  <span class="sidebar-brand-text">admix-kit docs</span>
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../prepare-dataset.html">Dataset preparation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../cli/index.html">Command line interface</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Command line interface</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/assoc-test.html">Association testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/genet-cor.html">Genetic correlation estimation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/pgs.html">Local ancestry-specific polygenic score</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/simulate-pheno.html">Phenotype simulation</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../api.html">Python API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Python API</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.assoc.marginal.html">admix.assoc.marginal</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.admix_grm.html">admix.cli.admix_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.admix_grm_merge.html">admix.cli.admix_grm_merge</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.genet_cor.html">admix.cli.genet_cor</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.calc_partial_pgs.html">admix.data.calc_partial_pgs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.Dataset.html">admix.Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.admix_geno.html">admix.simulate.admix_geno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno.html">admix.simulate.quant_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.binary_pheno.html">admix.simulate.binary_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno_1pop.html">admix.simulate.quant_pheno_1pop</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno_grm.html">admix.simulate.quant_pheno_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.allele_per_anc.html">admix.data.allele_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.af_per_anc.html">admix.data.af_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_dataset.html">admix.io.read_dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_lanc.html">admix.io.read_lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_rfmix.html">admix.io.read_rfmix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.tools.plink2.lift_over.html">admix.tools.plink2.lift_over</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.lanc.html">admix.plot.lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.admixture.html">admix.plot.admixture</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.manhattan.html">admix.plot.manhattan</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.qq.html">admix.plot.qq</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Example usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../simulate-admix-genotype.html">Admixed genotype simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/analyze-admix-simu-data.html">Basic statistics and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/assoc.html">Association testing (GWAS)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/genet-cor.html">Genetic correlation across local ancestry segments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/polygenic-scoring.html">Polygenic scoring for each local ancestry background</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/dataset.html">admix.Dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/select-admixed.html">Selecting admixed individuals</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <h1>Source code for admix.data._geno</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span> <span class="nn">dask.array</span> <span class="k">as</span> <span class="nn">da</span>
<span class="kn">import</span> <span class="nn">admix</span>
<span class="kn">import</span> <span class="nn">dask</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">List</span>
<span class="kn">import</span> <span class="nn">dapgen</span>


<span class="k">def</span> <span class="nf">calc_snp_prior_var</span><span class="p">(</span><span class="n">df_snp_info</span><span class="p">,</span> <span class="n">her_model</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the SNP prior variance from SNP information</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">her_model</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;uniform&quot;</span><span class="p">,</span> <span class="s2">&quot;gcta&quot;</span><span class="p">,</span> <span class="s2">&quot;ldak&quot;</span><span class="p">,</span> <span class="s2">&quot;mafukb&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">her_model</span> <span class="o">==</span> <span class="s2">&quot;uniform&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df_snp_info</span><span class="p">))</span>
    <span class="k">elif</span> <span class="n">her_model</span> <span class="o">==</span> <span class="s2">&quot;gcta&quot;</span><span class="p">:</span>
        <span class="n">freq</span> <span class="o">=</span> <span class="n">df_snp_info</span><span class="p">[</span><span class="s2">&quot;FREQ&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">freq</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span> <span class="s2">&quot;frequencies should be larger than zero&quot;</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">float_power</span><span class="p">(</span><span class="n">freq</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">freq</span><span class="p">),</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">her_model</span> <span class="o">==</span> <span class="s2">&quot;mafukb&quot;</span><span class="p">:</span>
        <span class="c1"># MAF-dependent genetic architecture, \alpha = -0.38 estimated from meta-analysis in UKB traits</span>
        <span class="n">freq</span> <span class="o">=</span> <span class="n">df_snp_info</span><span class="p">[</span><span class="s2">&quot;FREQ&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">freq</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span> <span class="s2">&quot;frequencies should be larger than zero&quot;</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">float_power</span><span class="p">(</span><span class="n">freq</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">freq</span><span class="p">),</span> <span class="o">-</span><span class="mf">0.38</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">her_model</span> <span class="o">==</span> <span class="s2">&quot;ldak&quot;</span><span class="p">:</span>
        <span class="n">freq</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="n">df_snp_info</span><span class="p">[</span><span class="s2">&quot;FREQ&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">df_snp_info</span><span class="p">[</span><span class="s2">&quot;LDAK_WEIGHT&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">float_power</span><span class="p">(</span><span class="n">freq</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">freq</span><span class="p">),</span> <span class="o">-</span><span class="mf">0.25</span><span class="p">)</span> <span class="o">*</span> <span class="n">weight</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span>


<span class="k">def</span> <span class="nf">impute_with_mean</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;impute the each entry using the mean of the input matrix np.mean(mat, axis=axis)</span>
<span class="sd">    axis = 1 corresponds to row-wise imputation</span>
<span class="sd">    axis = 0 corresponds to column-wise imputation</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    mat : np.ndarray</span>
<span class="sd">        input matrix. For reminder, the genotype matrix is with shape (n_snp, n_indiv)</span>
<span class="sd">    inplace : bool</span>
<span class="sd">        whether to return a new dataset or modify the input dataset</span>
<span class="sd">    axis : int</span>
<span class="sd">        axis to impute along</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    if inplace:</span>
<span class="sd">        mat : np.ndarray</span>
<span class="sd">            (n_snp, n_indiv) matrix</span>
<span class="sd">    else:</span>
<span class="sd">        None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">axis</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;axis should be 0 or 1&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">inplace</span><span class="p">:</span>
        <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1"># impute the missing genotypes with the mean of each row</span>
    <span class="n">mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="n">axis</span><span class="p">)</span>
    <span class="n">nanidx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">mat</span><span class="p">))</span>

    <span class="c1"># index the mean using the nanidx[1 - axis]</span>
    <span class="c1"># axis = 1, row-wise imputation, index the mean using the nanidx[0]</span>
    <span class="c1"># axis = 0, columnw-ise imputation, index the mean using the nanidx[1]</span>
    <span class="n">mat</span><span class="p">[</span><span class="n">nanidx</span><span class="p">]</span> <span class="o">=</span> <span class="n">mean</span><span class="p">[</span><span class="n">nanidx</span><span class="p">[</span><span class="mi">1</span> <span class="o">-</span> <span class="n">axis</span><span class="p">]]</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">inplace</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">mat</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>


<span class="k">def</span> <span class="nf">geno_mult_mat</span><span class="p">(</span>
    <span class="n">geno</span><span class="p">:</span> <span class="n">da</span><span class="o">.</span><span class="n">Array</span><span class="p">,</span>
    <span class="n">mat</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">impute_geno</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">mat_dim</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;snp&quot;</span><span class="p">,</span>
    <span class="n">return_snp_var</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Multiply genotype matrix with another matrix</span>

<span class="sd">    Chunk of genotype matrix will be read sequentially along the SNP dimension,</span>
<span class="sd">    and multiplied with the `mat`.</span>

<span class="sd">    Without transpose, result will be (n_snp, n_rep)</span>
<span class="sd">    With transpose, result will be (n_indiv, n_rep)</span>

<span class="sd">    Missing values in geno will be imputed with the mean of the genotype matrix.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    geno : da.Array</span>
<span class="sd">        Genotype matrix with shape (n_snp, n_indiv)</span>
<span class="sd">        geno.chunk contains the chunk of genotype matrix to be multiplied</span>
<span class="sd">    mat : np.ndarray</span>
<span class="sd">        Matrix to be multiplied with the genotype matrix. If the passed variable</span>
<span class="sd">        is a vector, it will be transformed to be a 1-column matrix.</span>
<span class="sd">    impute_geno : bool</span>
<span class="sd">        Whether to impute missing values with the mean of the genotype matrix</span>
<span class="sd">    mat_dim : str</span>
<span class="sd">        First dimension of the `mat`, either &quot;snp&quot; or &quot;indiv&quot;</span>
<span class="sd">        Whether to transpose the genotype matrix and calulate geno.T @ mat</span>
<span class="sd">    return_snp_var : bool</span>
<span class="sd">        Whether to return the variance of each SNP, useful in simple linear</span>
<span class="sd">        regression</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    np.ndarray</span>
<span class="sd">        Result of the multiplication</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">mat_dim</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;snp&quot;</span><span class="p">,</span> <span class="s2">&quot;indiv&quot;</span><span class="p">],</span> <span class="s2">&quot;mat_dim should be `snp` or `indiv`&quot;</span>

    <span class="k">if</span> <span class="n">mat</span><span class="o">.</span><span class="n">ndim</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">]</span>
    <span class="c1"># chunks over SNPs</span>
    <span class="n">chunks</span> <span class="o">=</span> <span class="n">geno</span><span class="o">.</span><span class="n">chunks</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">chunks</span><span class="p">),</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">n_snp</span><span class="p">,</span> <span class="n">n_indiv</span> <span class="o">=</span> <span class="n">geno</span><span class="o">.</span><span class="n">shape</span>
    <span class="n">n_rep</span> <span class="o">=</span> <span class="n">mat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

    <span class="n">snp_var</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_snp</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mat_dim</span> <span class="o">==</span> <span class="s2">&quot;indiv&quot;</span><span class="p">:</span>
        <span class="c1"># geno: (n_snp, n_indiv)</span>
        <span class="c1"># mat: (n_indiv, n_rep)</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">mat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">n_indiv</span>
        <span class="p">),</span> <span class="s2">&quot;when mat_dim is &#39;indiv&#39;, matrix should be of shape (n_indiv, n_rep)&quot;</span>
        <span class="n">ret</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_snp</span><span class="p">,</span> <span class="n">n_rep</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;admix.data.geno_mult_mat&quot;</span><span class="p">):</span>
            <span class="n">start</span><span class="p">,</span> <span class="n">stop</span> <span class="o">=</span> <span class="n">indices</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">indices</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
            <span class="n">geno_chunk</span> <span class="o">=</span> <span class="n">geno</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
            <span class="c1"># impute missing genotype</span>
            <span class="k">if</span> <span class="n">impute_geno</span><span class="p">:</span>
                <span class="n">impute_with_mean</span><span class="p">(</span><span class="n">geno_chunk</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">ret</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">geno_chunk</span><span class="p">,</span> <span class="n">mat</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">return_snp_var</span><span class="p">:</span>
                <span class="n">snp_var</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">geno_chunk</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">mat_dim</span> <span class="o">==</span> <span class="s2">&quot;snp&quot;</span><span class="p">:</span>
        <span class="c1"># geno: (n_indiv, n_snp)</span>
        <span class="c1"># mat: (n_snp, n_rep)</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">mat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">n_snp</span>
        <span class="p">),</span> <span class="s2">&quot;when mat_dim is &#39;snp&#39;, matrix should be of shape (n_snp, n_rep)&quot;</span>
        <span class="n">ret</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_indiv</span><span class="p">,</span> <span class="n">n_rep</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;admix.data.geno_mult_mat&quot;</span><span class="p">):</span>
            <span class="n">start</span><span class="p">,</span> <span class="n">stop</span> <span class="o">=</span> <span class="n">indices</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">indices</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
            <span class="n">geno_chunk</span> <span class="o">=</span> <span class="n">geno</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
            <span class="c1"># impute missing genotype</span>
            <span class="k">if</span> <span class="n">impute_geno</span><span class="p">:</span>
                <span class="n">impute_with_mean</span><span class="p">(</span><span class="n">geno_chunk</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">ret</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">geno_chunk</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">mat</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="p">:])</span>

            <span class="k">if</span> <span class="n">return_snp_var</span><span class="p">:</span>
                <span class="n">snp_var</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">geno_chunk</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;mat_dim should be `snp` or `indiv`&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">return_snp_var</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">ret</span><span class="p">,</span> <span class="n">snp_var</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">ret</span>


<span class="k">def</span> <span class="nf">grm</span><span class="p">(</span><span class="n">geno</span><span class="p">:</span> <span class="n">da</span><span class="o">.</span><span class="n">Array</span><span class="p">,</span> <span class="n">subpopu</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">std_method</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;std&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the GRM matrix</span>
<span class="sd">    This function is to serve as an alternative of GCTA --make-grm</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    geno: admix.Dataset</span>
<span class="sd">        genotype (n_snp, n_indiv) matrix</span>
<span class="sd">    subpopu : np.ndarray</span>
<span class="sd">        subpopulation labels, with shape (n_indiv,). The allele frequencies and</span>
<span class="sd">        normalization are performed separately within each subpopulation.</span>
<span class="sd">    std_method : str</span>
<span class="sd">        Method to standardize the GRM. Currently supported:</span>
<span class="sd">        &quot;std&quot; (standardize to have mean 0 and variance 1),</span>
<span class="sd">        &quot;allele&quot; (standardize to have mean 0 but no scaling)</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    np.ndarray</span>
<span class="sd">        GRM matrix (n_indiv, n_indiv)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">normalize_geno</span><span class="p">(</span><span class="n">g</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Normalize the genotype matrix&quot;&quot;&quot;</span>
        <span class="c1"># impute missing genotypes</span>
        <span class="n">g</span> <span class="o">=</span> <span class="n">impute_with_mean</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="c1"># normalize</span>
        <span class="k">if</span> <span class="n">std_method</span> <span class="o">==</span> <span class="s2">&quot;std&quot;</span><span class="p">:</span>
            <span class="n">g</span> <span class="o">=</span> <span class="p">(</span><span class="n">g</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)[:,</span> <span class="kc">None</span><span class="p">])</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)[:,</span> <span class="kc">None</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">std_method</span> <span class="o">==</span> <span class="s2">&quot;allele&quot;</span><span class="p">:</span>
            <span class="n">g</span> <span class="o">=</span> <span class="n">g</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)[:,</span> <span class="kc">None</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;std_method should be either `std` or `allele`&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">g</span>

    <span class="k">assert</span> <span class="n">std_method</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;std&quot;</span><span class="p">,</span> <span class="s2">&quot;allele&quot;</span><span class="p">],</span> <span class="s2">&quot;std_method should be `std` or `allele`&quot;</span>
    <span class="n">n_snp</span> <span class="o">=</span> <span class="n">geno</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">n_indiv</span> <span class="o">=</span> <span class="n">geno</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">subpopu</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">n_indiv</span> <span class="o">==</span> <span class="n">subpopu</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="p">),</span> <span class="s2">&quot;subpopu should have the same length as the number of individuals&quot;</span>
        <span class="n">unique_subpopu</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">subpopu</span><span class="p">)</span>
        <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_subpopu</span><span class="p">)</span><span class="si">}</span><span class="s2"> subpopulations found: </span><span class="si">{</span><span class="n">unique_subpopu</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Calculating GRM matrix with </span><span class="si">{</span><span class="n">n_snp</span><span class="si">}</span><span class="s2"> SNPs and </span><span class="si">{</span><span class="n">n_indiv</span><span class="si">}</span><span class="s2"> individuals&quot;</span>
    <span class="p">)</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">snp_chunks</span> <span class="o">=</span> <span class="n">geno</span><span class="o">.</span><span class="n">chunks</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">snp_chunks</span><span class="p">),</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;admix.data.grm&quot;</span><span class="p">):</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">stop</span> <span class="o">=</span> <span class="n">indices</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">indices</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">geno_chunk</span> <span class="o">=</span> <span class="n">geno</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">subpopu</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">popu</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">subpopu</span><span class="p">):</span>
                <span class="n">geno_chunk</span><span class="p">[:,</span> <span class="n">subpopu</span> <span class="o">==</span> <span class="n">popu</span><span class="p">]</span> <span class="o">=</span> <span class="n">normalize_geno</span><span class="p">(</span>
                    <span class="n">geno_chunk</span><span class="p">[:,</span> <span class="n">subpopu</span> <span class="o">==</span> <span class="n">popu</span><span class="p">]</span>
                <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">geno_chunk</span> <span class="o">=</span> <span class="n">normalize_geno</span><span class="p">(</span><span class="n">geno_chunk</span><span class="p">)</span>

        <span class="n">mat</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">geno_chunk</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">geno_chunk</span><span class="p">)</span> <span class="o">/</span> <span class="n">n_snp</span>
    <span class="k">return</span> <span class="n">mat</span>


<span class="k">def</span> <span class="nf">admix_grm</span><span class="p">(</span>
    <span class="n">geno</span><span class="p">:</span> <span class="n">da</span><span class="o">.</span><span class="n">Array</span><span class="p">,</span> <span class="n">lanc</span><span class="p">:</span> <span class="n">da</span><span class="o">.</span><span class="n">Array</span><span class="p">,</span> <span class="n">n_anc</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">snp_prior_var</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate ancestry specific GRM matrix</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    geno : da.Array</span>
<span class="sd">        Genotype matrix with shape (n_snp, n_indiv, 2)</span>
<span class="sd">    lanc : np.ndarray</span>
<span class="sd">        Local ancestry matrix with shape (n_snp, n_indiv, 2)</span>
<span class="sd">    n_anc : int</span>
<span class="sd">        Number of ancestral populations</span>
<span class="sd">    snp_prior_var : np.ndarray</span>
<span class="sd">        Prior variance of each SNP, shape (n_snp,)</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    G1: np.ndarray</span>
<span class="sd">        ancestry specific GRM matrix for the 1st ancestry</span>
<span class="sd">    G2: np.ndarray</span>
<span class="sd">        ancestry specific GRM matrix for the 2nd ancestry</span>
<span class="sd">    G12: np.ndarray</span>
<span class="sd">        ancestry specific GRM matrix for cross term of the 1st and 2nd ancestry</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">assert</span> <span class="n">n_anc</span> <span class="o">==</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;only two-way admixture is implemented&quot;</span>
    <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">geno</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">lanc</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

    <span class="n">apa</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">allele_per_anc</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span> <span class="n">lanc</span><span class="p">,</span> <span class="n">n_anc</span><span class="o">=</span><span class="n">n_anc</span><span class="p">)</span>
    <span class="n">n_snp</span><span class="p">,</span> <span class="n">n_indiv</span> <span class="o">=</span> <span class="n">apa</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">snp_prior_var</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">snp_prior_var</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">n_snp</span><span class="p">)</span>
    <span class="n">snp_prior_var_sum</span> <span class="o">=</span> <span class="n">snp_prior_var</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">G1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">([</span><span class="n">n_indiv</span><span class="p">,</span> <span class="n">n_indiv</span><span class="p">])</span>
    <span class="n">G2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">([</span><span class="n">n_indiv</span><span class="p">,</span> <span class="n">n_indiv</span><span class="p">])</span>
    <span class="n">G12</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">([</span><span class="n">n_indiv</span><span class="p">,</span> <span class="n">n_indiv</span><span class="p">])</span>

    <span class="n">snp_chunks</span> <span class="o">=</span> <span class="n">apa</span><span class="o">.</span><span class="n">chunks</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">snp_chunks</span><span class="p">),</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;admix.data.admix_grm&quot;</span><span class="p">):</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">stop</span> <span class="o">=</span> <span class="n">indices</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">indices</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">apa_chunk</span> <span class="o">=</span> <span class="n">apa</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>

        <span class="c1"># multiply by the prior variance on each SNP</span>
        <span class="n">apa_chunk</span> <span class="o">*=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">snp_prior_var</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">])[:,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span>
        <span class="n">a1_chunk</span><span class="p">,</span> <span class="n">a2_chunk</span> <span class="o">=</span> <span class="n">apa_chunk</span><span class="p">[:,</span> <span class="p">:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">apa_chunk</span><span class="p">[:,</span> <span class="p">:,</span> <span class="mi">1</span><span class="p">]</span>

        <span class="n">G1</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">a1_chunk</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">a1_chunk</span><span class="p">)</span> <span class="o">/</span> <span class="n">snp_prior_var_sum</span>
        <span class="n">G2</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">a2_chunk</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">a2_chunk</span><span class="p">)</span> <span class="o">/</span> <span class="n">snp_prior_var_sum</span>
        <span class="n">G12</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">a1_chunk</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">a2_chunk</span><span class="p">)</span> <span class="o">/</span> <span class="n">snp_prior_var_sum</span>

    <span class="k">return</span> <span class="n">G1</span><span class="p">,</span> <span class="n">G2</span><span class="p">,</span> <span class="n">G12</span>


<span class="k">def</span> <span class="nf">admix_grm_equal_var</span><span class="p">(</span>
    <span class="n">geno</span><span class="p">:</span> <span class="n">da</span><span class="o">.</span><span class="n">Array</span><span class="p">,</span> <span class="n">lanc</span><span class="p">:</span> <span class="n">da</span><span class="o">.</span><span class="n">Array</span><span class="p">,</span> <span class="n">n_anc</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">snp_prior_var</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate ancestry specific GRM matrix K1, K2 (assuming equal variances for ancestries)</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    geno : da.Array</span>
<span class="sd">        Genotype matrix with shape (n_snp, n_indiv, 2)</span>
<span class="sd">    lanc : np.ndarray</span>
<span class="sd">        Local ancestry matrix with shape (n_snp, n_indiv, 2)</span>
<span class="sd">    n_anc : int</span>
<span class="sd">        Number of ancestral populations</span>
<span class="sd">    snp_prior_var : np.ndarray</span>
<span class="sd">        Prior variance of each SNP, shape (n_snp,)</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    K1: np.ndarray</span>
<span class="sd">        sum of diagonal terms</span>
<span class="sd">    K2: np.ndarray</span>
<span class="sd">        off-diagonal terms</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">geno</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">lanc</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

    <span class="n">apa</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">allele_per_anc</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span> <span class="n">lanc</span><span class="p">,</span> <span class="n">n_anc</span><span class="o">=</span><span class="n">n_anc</span><span class="p">)</span>
    <span class="n">n_snp</span><span class="p">,</span> <span class="n">n_indiv</span> <span class="o">=</span> <span class="n">apa</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">snp_prior_var</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">snp_prior_var</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">n_snp</span><span class="p">)</span>
    <span class="n">snp_prior_var_sum</span> <span class="o">=</span> <span class="n">snp_prior_var</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

    <span class="n">K1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">([</span><span class="n">n_indiv</span><span class="p">,</span> <span class="n">n_indiv</span><span class="p">])</span>
    <span class="n">K2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">([</span><span class="n">n_indiv</span><span class="p">,</span> <span class="n">n_indiv</span><span class="p">])</span>

    <span class="n">snp_chunks</span> <span class="o">=</span> <span class="n">apa</span><span class="o">.</span><span class="n">chunks</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">snp_chunks</span><span class="p">),</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;admix.data.admix_grm_equal_var&quot;</span><span class="p">):</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">stop</span> <span class="o">=</span> <span class="n">indices</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">indices</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">apa_chunk</span> <span class="o">=</span> <span class="n">apa</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>

        <span class="c1"># multiply by the prior variance on each SNP</span>
        <span class="n">apa_chunk</span> <span class="o">*=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">snp_prior_var</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">])[:,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span>

        <span class="c1"># diagonal terms</span>
        <span class="k">for</span> <span class="n">i_anc</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_anc</span><span class="p">):</span>
            <span class="n">a_chunk</span> <span class="o">=</span> <span class="n">apa_chunk</span><span class="p">[:,</span> <span class="p">:,</span> <span class="n">i_anc</span><span class="p">]</span>
            <span class="n">K1</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">a_chunk</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">a_chunk</span><span class="p">)</span> <span class="o">/</span> <span class="n">snp_prior_var_sum</span>

        <span class="c1"># off-diagonal terms</span>
        <span class="k">for</span> <span class="n">i_anc</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_anc</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">j_anc</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i_anc</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">n_anc</span><span class="p">):</span>
                <span class="n">a1_chunk</span><span class="p">,</span> <span class="n">a2_chunk</span> <span class="o">=</span> <span class="n">apa_chunk</span><span class="p">[:,</span> <span class="p">:,</span> <span class="n">i_anc</span><span class="p">],</span> <span class="n">apa_chunk</span><span class="p">[:,</span> <span class="p">:,</span> <span class="n">j_anc</span><span class="p">]</span>
                <span class="n">K2</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">a1_chunk</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">a2_chunk</span><span class="p">)</span> <span class="o">/</span> <span class="n">snp_prior_var_sum</span>

    <span class="n">K2</span> <span class="o">=</span> <span class="n">K2</span> <span class="o">+</span> <span class="n">K2</span><span class="o">.</span><span class="n">T</span>
    <span class="k">return</span> <span class="n">K1</span><span class="p">,</span> <span class="n">K2</span>


<span class="k">def</span> <span class="nf">admix_ld</span><span class="p">(</span><span class="n">dset</span><span class="p">:</span> <span class="n">admix</span><span class="o">.</span><span class="n">Dataset</span><span class="p">,</span> <span class="n">cov</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate ancestry specific LD matrices</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    dset: admix.Dataset</span>
<span class="sd">        dataset containing geno, lanc</span>
<span class="sd">    cov : Optional[np.ndarray]</span>
<span class="sd">        (n_indiv, n_cov) covariates of the genotypes, an all `1` intercept covariate will always be added</span>
<span class="sd">        so there is no need to add the intercept in covariates.</span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    K1: np.ndarray</span>
<span class="sd">        ancestry specific LD matrix for the 1st ancestry</span>
<span class="sd">    K2: np.ndarray</span>
<span class="sd">        ancestry specific LD matrix for the 2nd ancestry</span>
<span class="sd">    K12: np.ndarray</span>
<span class="sd">        ancestry specific LD matrix for cross term of the 1st and 2nd ancestry</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">dset</span><span class="o">.</span><span class="n">n_anc</span> <span class="o">==</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;admix_ld only works for 2 ancestries for now&quot;</span>
    <span class="n">apa</span> <span class="o">=</span> <span class="n">dset</span><span class="o">.</span><span class="n">allele_per_anc</span><span class="p">()</span>

    <span class="n">n_snp</span><span class="p">,</span> <span class="n">n_indiv</span> <span class="o">=</span> <span class="n">apa</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>

    <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span> <span class="o">=</span> <span class="n">apa</span><span class="p">[:,</span> <span class="p">:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">apa</span><span class="p">[:,</span> <span class="p">:,</span> <span class="mi">1</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">cov</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">cov</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="n">n_indiv</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">cov</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="n">n_indiv</span><span class="p">,</span> <span class="mi">1</span><span class="p">)),</span> <span class="n">cov</span><span class="p">])</span>
    <span class="c1"># projection = I - X * (X&#39;X)^-1 * X&#39;</span>
    <span class="n">cov_proj_mat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="n">n_indiv</span><span class="p">)</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">multi_dot</span><span class="p">(</span>
        <span class="p">[</span><span class="n">cov</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">inv</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">cov</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">cov</span><span class="p">)),</span> <span class="n">cov</span><span class="o">.</span><span class="n">T</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">a1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span> <span class="n">cov_proj_mat</span><span class="p">)</span>
    <span class="n">a2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">a2</span><span class="p">,</span> <span class="n">cov_proj_mat</span><span class="p">)</span>
    <span class="c1"># center with row mean</span>
    <span class="c1"># a1 -= a1.mean(axis=1, keepdims=True)</span>
    <span class="c1"># a2 -= a2.mean(axis=1, keepdims=True)</span>
    <span class="n">ld1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span> <span class="n">a1</span><span class="o">.</span><span class="n">T</span><span class="p">)</span> <span class="o">/</span> <span class="n">n_indiv</span>
    <span class="n">ld2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">a2</span><span class="p">,</span> <span class="n">a2</span><span class="o">.</span><span class="n">T</span><span class="p">)</span> <span class="o">/</span> <span class="n">n_indiv</span>
    <span class="n">ld12</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="o">.</span><span class="n">T</span><span class="p">)</span> <span class="o">/</span> <span class="n">n_indiv</span>
    <span class="n">ld1</span><span class="p">,</span> <span class="n">ld2</span><span class="p">,</span> <span class="n">ld12</span> <span class="o">=</span> <span class="n">dask</span><span class="o">.</span><span class="n">compute</span><span class="p">(</span><span class="n">ld1</span><span class="p">,</span> <span class="n">ld2</span><span class="p">,</span> <span class="n">ld12</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;11&quot;</span><span class="p">:</span> <span class="n">ld1</span><span class="p">,</span> <span class="s2">&quot;22&quot;</span><span class="p">:</span> <span class="n">ld2</span><span class="p">,</span> <span class="s2">&quot;12&quot;</span><span class="p">:</span> <span class="n">ld12</span><span class="p">}</span>


<div class="viewcode-block" id="af_per_anc"><a class="viewcode-back" href="../../../_api/admix.data.af_per_anc.html#admix.data.af_per_anc">[docs]</a><span class="k">def</span> <span class="nf">af_per_anc</span><span class="p">(</span>
    <span class="n">geno</span><span class="p">,</span> <span class="n">lanc</span><span class="p">,</span> <span class="n">n_anc</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">return_nhaplo</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate allele frequency per ancestry</span>

<span class="sd">    If at one particular SNP locus, no SNP from one particular ancestry can be found</span>
<span class="sd">    the corresponding entries will be filled with np.NaN.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    geno: np.ndarray</span>
<span class="sd">        genotype matrix</span>
<span class="sd">    lanc: np.ndarray</span>
<span class="sd">        local ancestry matrix</span>
<span class="sd">    n_anc: int</span>
<span class="sd">        number of ancestries</span>
<span class="sd">    return_nhaplo: bool</span>
<span class="sd">        whether to return the number of haplotypes per ancestry</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    np.ndarray</span>
<span class="sd">        (n_snp, n_anc) length list of allele frequencies.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">geno</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">lanc</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="n">n_snp</span> <span class="o">=</span> <span class="n">geno</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">af</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_snp</span><span class="p">,</span> <span class="n">n_anc</span><span class="p">))</span>
    <span class="n">lanc_nhaplo</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_snp</span><span class="p">,</span> <span class="n">n_anc</span><span class="p">))</span>
    <span class="n">snp_chunks</span> <span class="o">=</span> <span class="n">geno</span><span class="o">.</span><span class="n">chunks</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">snp_chunks</span><span class="p">),</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;admix.data.af_per_anc&quot;</span><span class="p">):</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">stop</span> <span class="o">=</span> <span class="n">indices</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">indices</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
        <span class="n">geno_chunk</span> <span class="o">=</span> <span class="n">geno</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
        <span class="n">lanc_chunk</span> <span class="o">=</span> <span class="n">lanc</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">anc_i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_anc</span><span class="p">):</span>
            <span class="n">lanc_mask</span> <span class="o">=</span> <span class="n">lanc_chunk</span> <span class="o">==</span> <span class="n">anc_i</span>
            <span class="n">lanc_nhaplo</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="n">anc_i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">lanc_mask</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
            <span class="c1"># mask SNPs with local ancestry not `i_anc`</span>
            <span class="n">af</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="n">anc_i</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">masked_where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">logical_not</span><span class="p">(</span><span class="n">lanc_mask</span><span class="p">),</span> <span class="n">geno_chunk</span><span class="p">)</span>
                <span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
                <span class="o">.</span><span class="n">data</span>
            <span class="p">)</span> <span class="o">/</span> <span class="n">lanc_nhaplo</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="n">anc_i</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">return_nhaplo</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">af</span><span class="p">,</span> <span class="n">lanc_nhaplo</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">af</span></div>


<div class="viewcode-block" id="allele_per_anc"><a class="viewcode-back" href="../../../_api/admix.data.allele_per_anc.html#admix.data.allele_per_anc">[docs]</a><span class="k">def</span> <span class="nf">allele_per_anc</span><span class="p">(</span>
    <span class="n">geno</span><span class="p">:</span> <span class="n">da</span><span class="o">.</span><span class="n">Array</span><span class="p">,</span>
    <span class="n">lanc</span><span class="p">:</span> <span class="n">da</span><span class="o">.</span><span class="n">Array</span><span class="p">,</span>
    <span class="n">n_anc</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">center</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get allele count per ancestry</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    geno: da.Array</span>
<span class="sd">        genotype data</span>
<span class="sd">    lanc: da.Array</span>
<span class="sd">        local ancestry data</span>
<span class="sd">    n_anc: int</span>
<span class="sd">        number of ancestries</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Return allele counts per ancestries</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">center</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;center=True should not be used&quot;</span>
    <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">geno</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">lanc</span><span class="o">.</span><span class="n">shape</span><span class="p">),</span> <span class="s2">&quot;shape of `hap` and `lanc` are not equal&quot;</span>
    <span class="k">assert</span> <span class="n">geno</span><span class="o">.</span><span class="n">ndim</span> <span class="o">==</span> <span class="mi">3</span><span class="p">,</span> <span class="s2">&quot;`hap` and `lanc` should have three dimension&quot;</span>
    <span class="n">n_snp</span><span class="p">,</span> <span class="n">n_indiv</span><span class="p">,</span> <span class="n">n_haplo</span> <span class="o">=</span> <span class="n">geno</span><span class="o">.</span><span class="n">shape</span>
    <span class="k">assert</span> <span class="n">n_haplo</span> <span class="o">==</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;`n_haplo` should equal to 2, check your data&quot;</span>

    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span> <span class="n">da</span><span class="o">.</span><span class="n">Array</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">isinstance</span><span class="p">(</span>
        <span class="n">lanc</span><span class="p">,</span> <span class="n">da</span><span class="o">.</span><span class="n">Array</span>
    <span class="p">),</span> <span class="s2">&quot;`geno` and `lanc` should be dask array&quot;</span>

    <span class="c1"># make sure the chunk size along the ploidy axis to be 2</span>
    <span class="n">geno</span> <span class="o">=</span> <span class="n">geno</span><span class="o">.</span><span class="n">rechunk</span><span class="p">({</span><span class="mi">2</span><span class="p">:</span> <span class="mi">2</span><span class="p">})</span>
    <span class="n">lanc</span> <span class="o">=</span> <span class="n">lanc</span><span class="o">.</span><span class="n">rechunk</span><span class="p">({</span><span class="mi">2</span><span class="p">:</span> <span class="mi">2</span><span class="p">})</span>

    <span class="k">assert</span> <span class="p">(</span>
        <span class="n">geno</span><span class="o">.</span><span class="n">chunks</span> <span class="o">==</span> <span class="n">lanc</span><span class="o">.</span><span class="n">chunks</span>
    <span class="p">),</span> <span class="s2">&quot;`geno` and `lanc` should have the same chunk size&quot;</span>

    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">geno</span><span class="o">.</span><span class="n">chunks</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="p">(</span>
        <span class="s2">&quot;geno / lanc should not be chunked across the second dimension&quot;</span>
        <span class="s2">&quot;(individual dimension)&quot;</span>
    <span class="p">)</span>

    <span class="k">def</span> <span class="nf">helper</span><span class="p">(</span><span class="n">geno_chunk</span><span class="p">,</span> <span class="n">lanc_chunk</span><span class="p">,</span> <span class="n">n_anc</span><span class="p">):</span>
        <span class="n">n_snp</span><span class="p">,</span> <span class="n">n_indiv</span><span class="p">,</span> <span class="n">n_haplo</span> <span class="o">=</span> <span class="n">geno_chunk</span><span class="o">.</span><span class="n">shape</span>
        <span class="n">apa</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_snp</span><span class="p">,</span> <span class="n">n_indiv</span><span class="p">,</span> <span class="n">n_anc</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i_haplo</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_haplo</span><span class="p">):</span>
            <span class="n">haplo_hap</span> <span class="o">=</span> <span class="n">geno_chunk</span><span class="p">[:,</span> <span class="p">:,</span> <span class="n">i_haplo</span><span class="p">]</span>
            <span class="n">haplo_lanc</span> <span class="o">=</span> <span class="n">lanc_chunk</span><span class="p">[:,</span> <span class="p">:,</span> <span class="n">i_haplo</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">i_anc</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_anc</span><span class="p">):</span>
                <span class="n">apa</span><span class="p">[:,</span> <span class="p">:,</span> <span class="n">i_anc</span><span class="p">][</span><span class="n">haplo_lanc</span> <span class="o">==</span> <span class="n">i_anc</span><span class="p">]</span> <span class="o">+=</span> <span class="n">haplo_hap</span><span class="p">[</span><span class="n">haplo_lanc</span> <span class="o">==</span> <span class="n">i_anc</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">apa</span>

    <span class="c1"># the resulting chunk sizes will be the same as the input for snp, indiv</span>
    <span class="c1"># while the third dimension will be (n_anc, )</span>
    <span class="n">output_chunks</span> <span class="o">=</span> <span class="p">(</span><span class="n">geno</span><span class="o">.</span><span class="n">chunks</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">geno</span><span class="o">.</span><span class="n">chunks</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="p">(</span><span class="n">n_anc</span><span class="p">,))</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">da</span><span class="o">.</span><span class="n">map_blocks</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">geno_chunk</span><span class="p">,</span> <span class="n">lanc_chunk</span><span class="p">:</span> <span class="n">helper</span><span class="p">(</span>
            <span class="n">geno_chunk</span><span class="o">=</span><span class="n">geno_chunk</span><span class="p">,</span> <span class="n">lanc_chunk</span><span class="o">=</span><span class="n">lanc_chunk</span><span class="p">,</span> <span class="n">n_anc</span><span class="o">=</span><span class="n">n_anc</span>
        <span class="p">),</span>
        <span class="n">geno</span><span class="p">,</span>
        <span class="n">lanc</span><span class="p">,</span>
        <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span>
        <span class="n">chunks</span><span class="o">=</span><span class="n">output_chunks</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">res</span></div>


<span class="k">def</span> <span class="nf">calc_pgs</span><span class="p">(</span><span class="n">dset</span><span class="p">:</span> <span class="n">admix</span><span class="o">.</span><span class="n">Dataset</span><span class="p">,</span> <span class="n">df_weights</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">method</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate PGS for each individual</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    dset: admix.Dataset</span>
<span class="sd">        dataset object</span>
<span class="sd">    df_weights: pd.DataFrame</span>
<span class="sd">        weights for each individual</span>
<span class="sd">    method: str</span>
<span class="sd">        method to calculate PGS. Options are:</span>
<span class="sd">        - &quot;total&quot;: vanilla PGS</span>
<span class="sd">        - &quot;partial&quot;: partial PGS, calculate partial PGS for each local ancestry</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    np.ndarray</span>
<span class="sd">        PGS for each individual</span>
<span class="sd">        - method = &quot;total&quot;: (n_indiv, )</span>
<span class="sd">        - method = &quot;partial&quot;: (n_indiv, n_anc)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">method</span> <span class="ow">in</span> <span class="p">[</span>
        <span class="s2">&quot;total&quot;</span><span class="p">,</span>
        <span class="s2">&quot;partial&quot;</span><span class="p">,</span>
    <span class="p">],</span> <span class="s2">&quot;method should be either &#39;total&#39; or &#39;partial&#39;&quot;</span>
    <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
        <span class="n">dset</span><span class="o">.</span><span class="n">snp</span><span class="o">.</span><span class="n">index</span> <span class="o">==</span> <span class="n">df_weights</span><span class="o">.</span><span class="n">index</span>
    <span class="p">),</span> <span class="s2">&quot;`dset` and `df_weights` should have exactly the same index&quot;</span>

    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_weights</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;`df_weights` should have only one column&quot;</span>

    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;total&quot;</span><span class="p">:</span>
        <span class="n">pgs</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">geno_mult_mat</span><span class="p">(</span>
            <span class="n">dset</span><span class="o">.</span><span class="n">geno</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">2</span><span class="p">),</span> <span class="n">df_weights</span><span class="o">.</span><span class="n">values</span>
        <span class="p">)</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;partial&quot;</span><span class="p">:</span>
        <span class="n">n_anc</span> <span class="o">=</span> <span class="n">dset</span><span class="o">.</span><span class="n">n_anc</span>
        <span class="n">pgs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">dset</span><span class="o">.</span><span class="n">n_indiv</span><span class="p">,</span> <span class="n">n_anc</span><span class="p">))</span>
        <span class="n">apa</span> <span class="o">=</span> <span class="n">dset</span><span class="o">.</span><span class="n">allele_per_anc</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i_anc</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_anc</span><span class="p">):</span>
            <span class="n">pgs</span><span class="p">[:,</span> <span class="n">i_anc</span><span class="p">]</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">geno_mult_mat</span><span class="p">(</span>
                <span class="n">apa</span><span class="p">[:,</span> <span class="p">:,</span> <span class="n">i_anc</span><span class="p">],</span> <span class="n">df_weights</span><span class="o">.</span><span class="n">values</span>
            <span class="p">)</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;method should be either &#39;total&#39; or &#39;partial&#39;&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">pgs</span>


<div class="viewcode-block" id="calc_partial_pgs"><a class="viewcode-back" href="../../../_api/admix.data.calc_partial_pgs.html#admix.data.calc_partial_pgs">[docs]</a><span class="k">def</span> <span class="nf">calc_partial_pgs</span><span class="p">(</span>
    <span class="n">dset</span><span class="p">:</span> <span class="n">admix</span><span class="o">.</span><span class="n">Dataset</span><span class="p">,</span>
    <span class="n">df_weights</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">dset_ref</span><span class="p">:</span> <span class="n">admix</span><span class="o">.</span><span class="n">Dataset</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ref_pop_indiv</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">weight_col</span><span class="o">=</span><span class="s2">&quot;WEIGHT&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Given a vector of polygenic score weights, calculate polygenic scores with regard</span>
<span class="sd">    to every ancestry backgrounds for each individual.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    dset : admix.Dataset</span>
<span class="sd">        The admix.Dataset object.</span>
<span class="sd">    df_weights : pd.DataFrame</span>
<span class="sd">        The dataframe for polygenic score weights for each SNP, containing CHROM, POS, REF, ALT, WEIGHT, with index being SNP ID.</span>
<span class="sd">        CHROM, POS, REF, ALT will be aligned with `dset.snp`.</span>
<span class="sd">    dset_ref : admix.Dataset, optional</span>
<span class="sd">        The reference dataset object. Use `dapgen.align_snp` to align the SNPs between</span>
<span class="sd">        `dset` and `dset_ref`. `CHROM` and `POS` must match, with potential flips of</span>
<span class="sd">        `REF` and `ALT` allele coding.</span>
<span class="sd">    ref_pop_indiv : List[List[str]], optional</span>
<span class="sd">        The list of reference individual ID in `dset_ref`.</span>
<span class="sd">    weight_col : str, optional</span>
<span class="sd">        The column name for the weight in `df_weights`.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pd.DataFrame</span>
<span class="sd">        The polygenic scores (PGS) for each individual with shape of (n_indiv, n_anc).</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    AssertionError</span>
<span class="sd">        If both `dset_ref` and `ref_pop_indiv` are None or not None.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - The `dset` and `df_weights` should align, with potential allele flip.</span>
<span class="sd">    - If `dset_ref` is provided, `dset` and `dset_ref` should align, with potential allele flip.</span>
<span class="sd">    - The length of `ref_pop_indiv` should match with `dset.n_anc`.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="p">(</span><span class="n">dset_ref</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span>
        <span class="n">ref_pop_indiv</span> <span class="ow">is</span> <span class="kc">None</span>
    <span class="p">),</span> <span class="s2">&quot;both `dset_ref` and `ref_pop_indiv` should be None or not None&quot;</span>
    <span class="n">CALC_REF</span> <span class="o">=</span> <span class="n">dset_ref</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
    <span class="n">CHECK_COLS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;CHROM&quot;</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">,</span> <span class="s2">&quot;REF&quot;</span><span class="p">,</span> <span class="s2">&quot;ALT&quot;</span><span class="p">]</span>
    <span class="c1">## check input</span>
    <span class="n">idx1</span><span class="p">,</span> <span class="n">idx2</span><span class="p">,</span> <span class="n">sample_wgt_flip</span> <span class="o">=</span> <span class="n">dapgen</span><span class="o">.</span><span class="n">align_snp</span><span class="p">(</span>
        <span class="n">df1</span><span class="o">=</span><span class="n">dset</span><span class="o">.</span><span class="n">snp</span><span class="p">[</span><span class="n">CHECK_COLS</span><span class="p">],</span> <span class="n">df2</span><span class="o">=</span><span class="n">df_weights</span><span class="p">[</span><span class="n">CHECK_COLS</span><span class="p">]</span>
    <span class="p">)</span>

    <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">idx1</span> <span class="o">==</span> <span class="n">dset</span><span class="o">.</span><span class="n">snp</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
        <span class="n">idx2</span> <span class="o">==</span> <span class="n">df_weights</span><span class="o">.</span><span class="n">index</span>
    <span class="p">),</span> <span class="s2">&quot;`dset` and `df_weights` should align, with potential allele flip&quot;</span>

    <span class="k">if</span> <span class="n">CALC_REF</span><span class="p">:</span>
        <span class="n">idx1</span><span class="p">,</span> <span class="n">idx2</span><span class="p">,</span> <span class="n">ref_wgt_flip</span> <span class="o">=</span> <span class="n">dapgen</span><span class="o">.</span><span class="n">align_snp</span><span class="p">(</span>
            <span class="n">df1</span><span class="o">=</span><span class="n">dset</span><span class="o">.</span><span class="n">snp</span><span class="p">[</span><span class="n">CHECK_COLS</span><span class="p">],</span> <span class="n">df2</span><span class="o">=</span><span class="n">dset_ref</span><span class="o">.</span><span class="n">snp</span><span class="p">[</span><span class="n">CHECK_COLS</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">idx1</span> <span class="o">==</span> <span class="n">dset</span><span class="o">.</span><span class="n">snp</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
            <span class="n">idx2</span> <span class="o">==</span> <span class="n">dset_ref</span><span class="o">.</span><span class="n">snp</span><span class="o">.</span><span class="n">index</span>
        <span class="p">),</span> <span class="s2">&quot;`dset` and `dset_ref` should align, with potential allele flip&quot;</span>

    <span class="n">weights</span> <span class="o">=</span> <span class="n">df_weights</span><span class="p">[</span><span class="n">weight_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">sample_weights</span> <span class="o">=</span> <span class="n">weights</span> <span class="o">*</span> <span class="n">sample_wgt_flip</span>

    <span class="k">if</span> <span class="n">CALC_REF</span><span class="p">:</span>
        <span class="n">ref_weights</span> <span class="o">=</span> <span class="n">weights</span> <span class="o">*</span> <span class="n">ref_wgt_flip</span> <span class="o">*</span> <span class="n">sample_wgt_flip</span>

        <span class="k">assert</span> <span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">ref_pop_indiv</span><span class="p">)</span> <span class="o">==</span> <span class="n">dset</span><span class="o">.</span><span class="n">n_anc</span>
        <span class="p">),</span> <span class="s2">&quot;`len(ref_pops)` should match with `dset.n_anc`&quot;</span>

    <span class="c1">## scoring</span>
    <span class="n">dset_geno</span><span class="p">,</span> <span class="n">dset_lanc</span> <span class="o">=</span> <span class="n">dset</span><span class="o">.</span><span class="n">geno</span><span class="o">.</span><span class="n">compute</span><span class="p">(),</span> <span class="n">dset</span><span class="o">.</span><span class="n">lanc</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
    <span class="n">sample_pgs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">dset</span><span class="o">.</span><span class="n">n_indiv</span><span class="p">,</span> <span class="n">dset</span><span class="o">.</span><span class="n">n_anc</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">CALC_REF</span><span class="p">:</span>
        <span class="n">ref_geno_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">dset_ref</span><span class="p">[:,</span> <span class="n">pop</span><span class="p">]</span><span class="o">.</span><span class="n">geno</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">ref_pop_indiv</span><span class="p">]</span>
        <span class="n">ref_pgs</span> <span class="o">=</span> <span class="p">[[]</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">ref_pop_indiv</span><span class="p">]</span>
    <span class="c1"># iterate over each individuals</span>
    <span class="k">for</span> <span class="n">indiv_i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">dset</span><span class="o">.</span><span class="n">n_indiv</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;admix.data.calc_partial_pgs&quot;</span><span class="p">):</span>
        <span class="n">indiv_ref_pgs</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
        <span class="c1"># pgs for sample individuals</span>
        <span class="k">for</span> <span class="n">haplo_i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
            <span class="n">geno</span> <span class="o">=</span> <span class="n">dset_geno</span><span class="p">[:,</span> <span class="n">indiv_i</span><span class="p">,</span> <span class="n">haplo_i</span><span class="p">]</span>
            <span class="n">lanc</span> <span class="o">=</span> <span class="n">dset_lanc</span><span class="p">[:,</span> <span class="n">indiv_i</span><span class="p">,</span> <span class="n">haplo_i</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">lanc_i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">dset</span><span class="o">.</span><span class="n">n_anc</span><span class="p">):</span>
                <span class="c1"># sample</span>
                <span class="n">sample_pgs</span><span class="p">[</span><span class="n">indiv_i</span><span class="p">,</span> <span class="n">lanc_i</span><span class="p">]</span> <span class="o">+=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span>
                    <span class="n">geno</span><span class="p">[</span><span class="n">lanc</span> <span class="o">==</span> <span class="n">lanc_i</span><span class="p">],</span> <span class="n">sample_weights</span><span class="p">[</span><span class="n">lanc</span> <span class="o">==</span> <span class="n">lanc_i</span><span class="p">]</span>
                <span class="p">)</span>

                <span class="c1"># pgs for reference individuals</span>
                <span class="k">if</span> <span class="n">CALC_REF</span><span class="p">:</span>
                    <span class="n">ref_geno</span> <span class="o">=</span> <span class="n">ref_geno_list</span><span class="p">[</span><span class="n">lanc_i</span><span class="p">][</span><span class="n">lanc</span> <span class="o">==</span> <span class="n">lanc_i</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span>
                    <span class="k">if</span> <span class="n">ref_geno</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">ref_geno</span> <span class="o">=</span> <span class="n">ref_geno</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">ref_geno</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>
                        <span class="n">s</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">ref_weights</span><span class="p">[</span><span class="n">lanc</span> <span class="o">==</span> <span class="n">lanc_i</span><span class="p">],</span> <span class="n">ref_geno</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">s</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">ref_geno</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span>
                    <span class="n">indiv_ref_pgs</span><span class="p">[</span><span class="n">lanc_i</span><span class="p">]</span> <span class="o">+=</span> <span class="n">s</span>
        <span class="k">if</span> <span class="n">CALC_REF</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">lanc_i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">dset</span><span class="o">.</span><span class="n">n_anc</span><span class="p">):</span>
                <span class="n">ref_pgs</span><span class="p">[</span><span class="n">lanc_i</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">indiv_ref_pgs</span><span class="p">[</span><span class="n">lanc_i</span><span class="p">])</span>

    <span class="c1"># format ref_pgs: for each ancestry, we have n_indiv x (n_ref_indiv x 2)</span>
    <span class="c1"># each reference has 2 haplotypes</span>
    <span class="k">if</span> <span class="n">CALC_REF</span><span class="p">:</span>
        <span class="n">ref_pgs</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="n">data</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">(</span><span class="n">ref_pgs</span><span class="p">[</span><span class="n">i</span><span class="p">]),</span>
                <span class="n">index</span><span class="o">=</span><span class="n">dset</span><span class="o">.</span><span class="n">indiv</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
                <span class="n">columns</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span>
                    <span class="p">[[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_1&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_2&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">ref_pop_indiv</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">dset</span><span class="o">.</span><span class="n">n_anc</span><span class="p">)</span>
        <span class="p">]</span>
    <span class="n">sample_pgs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="n">data</span><span class="o">=</span><span class="n">sample_pgs</span><span class="p">,</span>
        <span class="n">index</span><span class="o">=</span><span class="n">dset</span><span class="o">.</span><span class="n">indiv</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;ANC</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">dset</span><span class="o">.</span><span class="n">n_anc</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)],</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="n">CALC_REF</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">sample_pgs</span><span class="p">,</span> <span class="n">ref_pgs</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">sample_pgs</span></div>
</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021-2024, Kangcheng Hou
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/kangchenghou/admix-kit" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js?v=b3ba4146"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../../../_static/scripts/furo.js?v=32e29ea5"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/tabs.js?v=3ee01567"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>