<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2024.01.29 -->
        <title>admix.cli._assoc - admix-kit docs</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/tabs.css?v=4c969af8" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">admix-kit docs</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  
  
  <span class="sidebar-brand-text">admix-kit docs</span>
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../prepare-dataset.html">Dataset preparation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../cli/index.html">Command line interface</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Command line interface</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/assoc-test.html">Association testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/genet-cor.html">Genetic correlation estimation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/pgs.html">Local ancestry-specific polygenic score</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/simulate-pheno.html">Phenotype simulation</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../api.html">Python API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Python API</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.assoc.marginal.html">admix.assoc.marginal</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.admix_grm.html">admix.cli.admix_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.admix_grm_merge.html">admix.cli.admix_grm_merge</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.genet_cor.html">admix.cli.genet_cor</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.calc_partial_pgs.html">admix.data.calc_partial_pgs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.Dataset.html">admix.Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.admix_geno.html">admix.simulate.admix_geno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno.html">admix.simulate.quant_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.binary_pheno.html">admix.simulate.binary_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno_1pop.html">admix.simulate.quant_pheno_1pop</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno_grm.html">admix.simulate.quant_pheno_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.allele_per_anc.html">admix.data.allele_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.af_per_anc.html">admix.data.af_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_dataset.html">admix.io.read_dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_lanc.html">admix.io.read_lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_rfmix.html">admix.io.read_rfmix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.tools.plink2.lift_over.html">admix.tools.plink2.lift_over</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.lanc.html">admix.plot.lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.admixture.html">admix.plot.admixture</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.manhattan.html">admix.plot.manhattan</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.qq.html">admix.plot.qq</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Example usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../simulate-admix-genotype.html">Admixed genotype simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/analyze-admix-simu-data.html">Basic statistics and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/assoc.html">Association testing (GWAS)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/genet-cor.html">Genetic correlation across local ancestry segments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/polygenic-scoring.html">Polygenic scoring for each local ancestry background</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/dataset.html">admix.Dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/select-admixed.html">Selecting admixed individuals</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <h1>Source code for admix.cli._assoc</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">admix</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Union</span><span class="p">,</span> <span class="n">List</span>
<span class="kn">from</span> <span class="nn">admix</span> <span class="kn">import</span> <span class="n">logger</span>
<span class="kn">from</span> <span class="nn">._utils</span> <span class="kn">import</span> <span class="n">log_params</span>


<div class="viewcode-block" id="assoc"><a class="viewcode-back" href="../../../cli/assoc-test.html#admix.cli.assoc">[docs]</a><span class="k">def</span> <span class="nf">assoc</span><span class="p">(</span>
    <span class="n">pfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">pheno</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">out</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">method</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="s2">&quot;ATT&quot;</span><span class="p">,</span>
    <span class="n">family</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;quant&quot;</span><span class="p">,</span>
    <span class="n">quantile_normalize</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">snp_list</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">fast</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Perform association testing.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    pfile : str</span>
<span class="sd">        Prefix to the PLINK2 file (.pgen should not be added). When using a method requiring</span>
<span class="sd">        local ancestry, a matching :code:`&lt;pfile&gt;.lanc` file should also exist.</span>
<span class="sd">    pheno : str</span>
<span class="sd">        Path to the phenotype file. The text file should be space delimited with header</span>
<span class="sd">        and one individual per row. 1st column: individual ID. 2nd column: phenotype</span>
<span class="sd">        values. 3rd - nth columns: covariates. NaN should be encoded as &quot;NA&quot; and these</span>
<span class="sd">        individuals will be removed in the analysis.</span>
<span class="sd">        Binary phenotype should be encoded as 0 and 1, and</span>
<span class="sd">        :code:`--family binary` should be used.  All</span>
<span class="sd">        columns will be used for the analysis. NaN should be encoded as &quot;NA&quot; and NaN</span>
<span class="sd">        will be imputed with the mean of each covariate. Categorical covariates will be</span>
<span class="sd">        converted to one hot encodings internally.</span>
<span class="sd">    out : str</span>
<span class="sd">        Path the output file. :code:`&lt;out&gt;.&lt;method&gt;.assoc` will be created.</span>
<span class="sd">    method : Union[str, List[str]]</span>
<span class="sd">        Method to use for association analysis (default ATT). Other methods include:</span>
<span class="sd">        TRACTOR, ADM, SNP1, HET</span>
<span class="sd">    family : str</span>
<span class="sd">        Family to use for association analysis (default quant). One of :code:`quant` or</span>
<span class="sd">        :code:`binary`.</span>
<span class="sd">    quantile_normalize : bool</span>
<span class="sd">        Whether to quantile normalize the phenotype and every covariate. When</span>
<span class="sd">        :code:`--family binary` is used, quantile normalization will only be applied</span>
<span class="sd">        to covariates.</span>
<span class="sd">    snp_list : str</span>
<span class="sd">        Path to the SNP list file. Each line should be a SNP ID. Only SNPs in the</span>
<span class="sd">        list will be used for the analysis.</span>
<span class="sd">    fast : bool</span>
<span class="sd">        Whether to use fast mode (default True).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">log_params</span><span class="p">(</span><span class="s2">&quot;assoc&quot;</span><span class="p">,</span> <span class="nb">locals</span><span class="p">())</span>
    <span class="k">assert</span> <span class="n">family</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;quant&quot;</span><span class="p">,</span> <span class="s2">&quot;binary&quot;</span><span class="p">],</span> <span class="s2">&quot;family must be either quant or binary&quot;</span>

    <span class="c1"># TODO: infer block size using memory use in dask-pgen read</span>
    <span class="n">dset</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read_dataset</span><span class="p">(</span><span class="n">pfile</span><span class="p">)</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">dset</span><span class="o">.</span><span class="n">n_snp</span><span class="si">}</span><span class="s2"> SNPs and </span><span class="si">{</span><span class="n">dset</span><span class="o">.</span><span class="n">n_indiv</span><span class="si">}</span><span class="s2"> individuals are loaded&quot;</span><span class="p">)</span>
    <span class="n">df_pheno</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">pheno</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">df_pheno</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">df_pheno</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">pheno_col</span> <span class="o">=</span> <span class="n">df_pheno</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">covar_cols</span> <span class="o">=</span> <span class="n">df_pheno</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">covar_cols</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">covar_cols</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;trait column: </span><span class="si">{</span><span class="n">pheno_col</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;covariate columns: </span><span class="si">{</span><span class="n">covar_cols</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">dset</span><span class="o">.</span><span class="n">append_indiv_info</span><span class="p">(</span><span class="n">df_pheno</span><span class="p">,</span> <span class="n">force_update</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># retain only individuals with non-missing phenotype,</span>
    <span class="c1"># or with non-completely missing covariate</span>
    <span class="n">indiv_mask</span> <span class="o">=</span> <span class="o">~</span><span class="n">dset</span><span class="o">.</span><span class="n">indiv</span><span class="p">[</span><span class="n">pheno_col</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">values</span>
    <span class="k">if</span> <span class="n">covar_cols</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">covar_mask</span> <span class="o">=</span> <span class="o">~</span><span class="p">(</span><span class="n">dset</span><span class="o">.</span><span class="n">indiv</span><span class="p">[</span><span class="n">covar_cols</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
        <span class="n">indiv_mask</span> <span class="o">&amp;=</span> <span class="n">covar_mask</span>
    <span class="n">dset</span> <span class="o">=</span> <span class="n">dset</span><span class="p">[:,</span> <span class="n">indiv_mask</span><span class="p">]</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">dset</span><span class="o">.</span><span class="n">n_indiv</span><span class="si">}</span><span class="s2"> individuals left &quot;</span>
        <span class="s2">&quot;after filtering for missing phenotype, or completely missing covariate&quot;</span>
    <span class="p">)</span>

    <span class="c1"># filter for SNPs</span>
    <span class="k">if</span> <span class="n">snp_list</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">snp_list</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">filter_snp_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">]</span>
        <span class="n">n_filter_snp</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">filter_snp_list</span><span class="p">)</span>
        <span class="n">filter_snp_list</span> <span class="o">=</span> <span class="n">dset</span><span class="o">.</span><span class="n">snp</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">dset</span><span class="o">.</span><span class="n">snp</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">filter_snp_list</span><span class="p">)]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filter_snp_list</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">n_filter_snp</span><span class="p">:</span>
            <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">n_filter_snp</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">filter_snp_list</span><span class="p">)</span><span class="si">}</span><span class="s2"> SNPs in </span><span class="si">{</span><span class="n">snp_list</span><span class="si">}</span><span class="s2"> are not in the dataset&quot;</span>
            <span class="p">)</span>
        <span class="n">dset</span> <span class="o">=</span> <span class="n">dset</span><span class="p">[</span><span class="n">filter_snp_list</span><span class="p">]</span>

    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">dset</span><span class="o">.</span><span class="n">n_snp</span><span class="si">}</span><span class="s2"> SNPs and </span><span class="si">{</span><span class="n">dset</span><span class="o">.</span><span class="n">n_indiv</span><span class="si">}</span><span class="s2"> individuals in the analysis&quot;</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">method</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">method</span> <span class="o">=</span> <span class="p">[</span><span class="n">method</span><span class="p">]</span>

    <span class="c1"># extract pheno_values and covar_values</span>
    <span class="n">pheno_values</span> <span class="o">=</span> <span class="n">dset</span><span class="o">.</span><span class="n">indiv</span><span class="p">[</span><span class="n">pheno_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="k">if</span> <span class="n">covar_cols</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">covar_values</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">convert_dummy</span><span class="p">(</span><span class="n">dset</span><span class="o">.</span><span class="n">indiv</span><span class="p">[</span><span class="n">covar_cols</span><span class="p">])</span><span class="o">.</span><span class="n">values</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">covar_values</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">quantile_normalize</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">family</span> <span class="o">==</span> <span class="s2">&quot;quant&quot;</span><span class="p">:</span>
            <span class="n">pheno_values</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">quantile_normalize</span><span class="p">(</span><span class="n">pheno_values</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">covar_values</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">covar_values</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
                <span class="n">covar_values</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">quantile_normalize</span><span class="p">(</span><span class="n">covar_values</span><span class="p">[:,</span> <span class="n">i</span><span class="p">])</span>

    <span class="n">dict_rls</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">method</span><span class="p">:</span>
        <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Performing association analysis with method </span><span class="si">{</span><span class="n">m</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">dict_rls</span><span class="p">[</span><span class="n">m</span><span class="p">]</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">assoc</span><span class="o">.</span><span class="n">marginal</span><span class="p">(</span>
            <span class="n">dset</span><span class="o">=</span><span class="n">dset</span><span class="p">,</span>
            <span class="n">pheno</span><span class="o">=</span><span class="n">pheno_values</span><span class="p">,</span>
            <span class="n">cov</span><span class="o">=</span><span class="n">covar_values</span><span class="p">,</span>
            <span class="n">method</span><span class="o">=</span><span class="n">m</span><span class="p">,</span>
            <span class="n">family</span><span class="o">=</span><span class="s2">&quot;logistic&quot;</span> <span class="k">if</span> <span class="n">family</span> <span class="o">==</span> <span class="s2">&quot;binary&quot;</span> <span class="k">else</span> <span class="s2">&quot;linear&quot;</span><span class="p">,</span>
            <span class="n">fast</span><span class="o">=</span><span class="n">fast</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">dict_rls</span><span class="p">:</span>
        <span class="n">dict_rls</span><span class="p">[</span><span class="n">m</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">m</span><span class="si">}</span><span class="s2">.assoc&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">float_format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%.6g</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">na_rep</span><span class="o">=</span><span class="s2">&quot;NA&quot;</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Output written to </span><span class="si">{</span><span class="n">out</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">m</span><span class="si">}</span><span class="s2">.assoc&quot;</span><span class="p">)</span></div>
</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021-2024, Kangcheng Hou
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/kangchenghou/admix-kit" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js?v=b3ba4146"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../../../_static/scripts/furo.js?v=32e29ea5"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/tabs.js?v=3ee01567"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>