<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2024.01.29 -->
        <title>admix.cli._utils - admix-kit docs</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/tabs.css?v=4c969af8" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">admix-kit docs</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  
  
  <span class="sidebar-brand-text">admix-kit docs</span>
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../prepare-dataset.html">Dataset preparation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../cli/index.html">Command line interface</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Command line interface</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/assoc-test.html">Association testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/genet-cor.html">Genetic correlation estimation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/pgs.html">Local ancestry-specific polygenic score</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/simulate-pheno.html">Phenotype simulation</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../api.html">Python API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Python API</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.assoc.marginal.html">admix.assoc.marginal</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.admix_grm.html">admix.cli.admix_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.admix_grm_merge.html">admix.cli.admix_grm_merge</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.genet_cor.html">admix.cli.genet_cor</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.calc_partial_pgs.html">admix.data.calc_partial_pgs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.Dataset.html">admix.Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.admix_geno.html">admix.simulate.admix_geno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno.html">admix.simulate.quant_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.binary_pheno.html">admix.simulate.binary_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno_1pop.html">admix.simulate.quant_pheno_1pop</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno_grm.html">admix.simulate.quant_pheno_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.allele_per_anc.html">admix.data.allele_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.af_per_anc.html">admix.data.af_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_dataset.html">admix.io.read_dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_lanc.html">admix.io.read_lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_rfmix.html">admix.io.read_rfmix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.tools.plink2.lift_over.html">admix.tools.plink2.lift_over</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.lanc.html">admix.plot.lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.admixture.html">admix.plot.admixture</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.manhattan.html">admix.plot.manhattan</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.qq.html">admix.plot.qq</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Example usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../simulate-admix-genotype.html">Admixed genotype simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/analyze-admix-simu-data.html">Basic statistics and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/assoc.html">Association testing (GWAS)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/genet-cor.html">Genetic correlation across local ancestry segments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/polygenic-scoring.html">Polygenic scoring for each local ancestry background</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/dataset.html">admix.Dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/select-admixed.html">Selecting admixed individuals</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <h1>Source code for admix.cli._utils</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">admix</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">dapgen</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Tuple</span>


<span class="k">def</span> <span class="nf">log_params</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">params</span><span class="p">):</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Received parameters: </span><span class="se">\n</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="se">\n</span><span class="s2">  &quot;</span>
        <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">  &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">=</span><span class="si">{</span><span class="n">v</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">params</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>
    <span class="p">)</span>


<span class="k">def</span> <span class="nf">_process_sample_map</span><span class="p">(</span><span class="n">root_dir</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Download and format sample map from 1000 Genomes&quot;&quot;&quot;</span>

    <span class="n">sample_map</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/pgen/all_chr.psam&quot;</span><span class="p">,</span>
        <span class="n">delim_whitespace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">unrelated_id</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/pgen/king.cutoff.out.id&quot;</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/metadata&quot;</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">sample_map</span><span class="p">[[</span><span class="s2">&quot;#IID&quot;</span><span class="p">,</span> <span class="s2">&quot;Population&quot;</span><span class="p">,</span> <span class="s2">&quot;SuperPop&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/metadata/full_sample.tsv&quot;</span><span class="p">,</span>
        <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># filter unrelated</span>
    <span class="n">unrelated_sample_map</span> <span class="o">=</span> <span class="n">sample_map</span><span class="p">[</span><span class="o">~</span><span class="n">sample_map</span><span class="p">[</span><span class="s2">&quot;#IID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">unrelated_id</span><span class="p">[</span><span class="s2">&quot;#IID&quot;</span><span class="p">])]</span>
    <span class="n">unrelated_sample_map</span><span class="p">[[</span><span class="s2">&quot;#IID&quot;</span><span class="p">,</span> <span class="s2">&quot;Population&quot;</span><span class="p">,</span> <span class="s2">&quot;SuperPop&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/metadata/unrelated_sample.tsv&quot;</span><span class="p">,</span>
        <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Population in unrelated sample map:&quot;</span><span class="p">)</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">unrelated_sample_map</span><span class="p">[</span><span class="s2">&quot;Population&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">())</span>


<span class="k">def</span> <span class="nf">_process_genetic_map</span><span class="p">(</span><span class="n">root_dir</span><span class="p">,</span> <span class="n">build</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Download and format genetic map</span>
<span class="sd">    1. call bash script to download genetic map to out/metadata/genetic_map/raw</span>
<span class="sd">    2. process the genetic map and save to out/metadata/genetic_map</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">assert</span> <span class="n">build</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">],</span> <span class="s2">&quot;build should be hg38 or hg19&quot;</span>

    <span class="k">if</span> <span class="n">build</span> <span class="o">==</span> <span class="s2">&quot;hg38&quot;</span><span class="p">:</span>
        <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;GRCh38&quot;</span>
    <span class="k">elif</span> <span class="n">build</span> <span class="o">==</span> <span class="s2">&quot;hg19&quot;</span><span class="p">:</span>
        <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;GRCh37&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;build should be hg38 or hg19&quot;</span><span class="p">)</span>

    <span class="n">cmds</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        mkdir -p </span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/metadata/genetic_map/raw &amp;&amp; cd </span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/metadata/genetic_map/raw</span>
<span class="s2">        wget https://bochet.gcc.biostat.washington.edu/beagle/genetic_maps/plink.</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">.map.zip</span>
<span class="s2">        unzip plink.</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">.map.zip</span>
<span class="s2">    &quot;&quot;&quot;</span>

    <span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">(</span><span class="n">cmds</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">):</span>
        <span class="c1"># move raw files to out/metadata/genetic_map</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/metadata/genetic_map/raw/plink.chr</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">.map&quot;</span><span class="p">,</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/metadata/genetic_map/chr</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">.map&quot;</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="c1"># clean up</span>
    <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/metadata/genetic_map/raw&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="kc">False</span><span class="p">:</span>
        <span class="c1"># OLD Eagle2 genetic map</span>
        <span class="n">raw_map_path</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">get_cache_data</span><span class="p">(</span><span class="s2">&quot;genetic_map&quot;</span><span class="p">,</span> <span class="n">build</span><span class="o">=</span><span class="n">build</span><span class="p">)</span>

        <span class="n">raw_map</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">raw_map_path</span><span class="p">,</span>
            <span class="n">delim_whitespace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/metadata/genetic_map&quot;</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">):</span>
            <span class="n">chrom_map</span> <span class="o">=</span> <span class="n">raw_map</span><span class="p">[</span><span class="n">raw_map</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">]]</span>
            <span class="n">chrom_map</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">root_dir</span><span class="si">}</span><span class="s2">/metadata/genetic_map/chr</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">.tsv&quot;</span><span class="p">,</span>
                <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span>


<div class="viewcode-block" id="get_1kg_ref"><a class="viewcode-back" href="../../../cli/misc.html#admix.cli.get_1kg_ref">[docs]</a><span class="k">def</span> <span class="nf">get_1kg_ref</span><span class="p">(</span>
    <span class="nb">dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">build</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">memory</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">16000</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the 1,000 reference genome in plink2 format.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    dir : str</span>
<span class="sd">        Directory where the reference genome is stored. The following files will be downloaded</span>
<span class="sd">        and stored in this directory:</span>
<span class="sd">        - ${dir}/pgen: containing plink2 files</span>
<span class="sd">        - ${dir}/metadata: containing metadata files</span>
<span class="sd">    build : str</span>
<span class="sd">        Build of the reference genome. hg38 or hg19.</span>
<span class="sd">    verbose : bool</span>
<span class="sd">        Whether to print out the progress of the download.</span>
<span class="sd">    memory : int</span>
<span class="sd">        Memory in MB to use for PLINK command. Default is 16000 (~16GB).</span>

<span class="sd">    Examples</span>
<span class="sd">    --------</span>
<span class="sd">    .. code-block:: bash</span>

<span class="sd">        admix get-1kg-ref --dir=/path/to/1kg --build hg38</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">log_params</span><span class="p">(</span><span class="s2">&quot;get-1kg-ref&quot;</span><span class="p">,</span> <span class="nb">locals</span><span class="p">())</span>

    <span class="k">def</span> <span class="nf">call_helper</span><span class="p">(</span><span class="n">cmd</span><span class="p">):</span>
        <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;$ </span><span class="si">{</span><span class="n">cmd</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
                <span class="n">cmd</span><span class="p">,</span>
                <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">DEVNULL</span><span class="p">,</span>
                <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">STDOUT</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="c1"># make sure plink2 is in path</span>
    <span class="n">plink2</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">get_dependency</span><span class="p">(</span><span class="s2">&quot;plink2&quot;</span><span class="p">)</span>

    <span class="c1"># step0: download metadata</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span> <span class="s2">&quot;metadata&quot;</span><span class="p">),</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Retrieving meta data...&quot;</span><span class="p">)</span>

    <span class="c1"># step1: download pgen</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span> <span class="s2">&quot;pgen&quot;</span><span class="p">),</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">build</span> <span class="o">==</span> <span class="s2">&quot;hg38&quot;</span><span class="p">:</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span>
            <span class="sa">f</span><span class="s2">&quot;mkdir -p </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen &amp;&amp;&quot;</span><span class="p">,</span>
            <span class="sa">f</span><span class="s2">&quot;wget https://www.dropbox.com/s/j72j6uciq5zuzii/all_hg38.pgen.zst?dl=1 -O </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw.pgen.zst &amp;&amp;&quot;</span><span class="p">,</span>
            <span class="sa">f</span><span class="s2">&quot;wget https://www.dropbox.com/s/ngbo2xm5ojw9koy/all_hg38_noannot.pvar.zst?dl=1 -O </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw.pvar.zst &amp;&amp;&quot;</span><span class="p">,</span>
            <span class="sa">f</span><span class="s2">&quot;wget https://www.dropbox.com/s/2e87z6nc4qexjjm/hg38_corrected.psam?dl=1 -O </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw.psam &amp;&amp;&quot;</span><span class="p">,</span>
            <span class="sa">f</span><span class="s2">&quot;wget https://www.dropbox.com/s/4zhmxpk5oclfplp/deg2_hg38.king.cutoff.out.id?dl=1 -O </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/king.cutoff.out.id&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="k">elif</span> <span class="n">build</span> <span class="o">==</span> <span class="s2">&quot;hg19&quot;</span><span class="p">:</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span>
            <span class="sa">f</span><span class="s2">&quot;mkdir -p </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen &amp;&amp;&quot;</span><span class="p">,</span>
            <span class="sa">f</span><span class="s2">&quot;wget https://www.dropbox.com/s/dps1kvlq338ukz8/all_phase3_ns.pgen.zst?dl=1 -O </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw.pgen.zst &amp;&amp;&quot;</span><span class="p">,</span>
            <span class="sa">f</span><span class="s2">&quot;wget https://www.dropbox.com/s/uqk3gfhwsvf7bf3/all_phase3_ns_noannot.pvar.zst?dl=1 -O </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw.pvar.zst &amp;&amp;&quot;</span><span class="p">,</span>
            <span class="sa">f</span><span class="s2">&quot;wget https://www.dropbox.com/s/6ppo144ikdzery5/phase3_corrected.psam?dl=1 -O </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw.psam &amp;&amp;&quot;</span><span class="p">,</span>
            <span class="sa">f</span><span class="s2">&quot;wget https://www.dropbox.com/s/zj8d14vv9mp6x3c/deg2_phase3.king.cutoff.out.id?dl=1 -O </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/king.cutoff.out.id&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unknown build: </span><span class="si">{</span><span class="n">build</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># assume that ${dir}/pgen already contains raw.pgen / raw.pvar / raw.psam / king.cutoff.out.id</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;raw.pgen.zst&quot;</span><span class="p">,</span> <span class="s2">&quot;raw.pvar.zst&quot;</span><span class="p">,</span> <span class="s2">&quot;raw.psam&quot;</span><span class="p">]</span>
        <span class="p">]</span>
    <span class="p">):</span>
        <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Please download the reference genome to </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen using the following commands:</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="o">+</span> <span class="s2">&quot;$ &quot;</span>
            <span class="o">+</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmds</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">return</span>

    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Decompressing plink files&quot;</span><span class="p">)</span>
    <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">plink2</span><span class="si">}</span><span class="s2"> --zst-decompress </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw.pgen.zst &gt; </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw.pgen &amp;&amp;&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">plink2</span><span class="si">}</span><span class="s2"> --zst-decompress </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw.pvar.zst &gt; </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw.pvar&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">call_helper</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmds</span><span class="p">))</span>

    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Basic QCing: bi-allelic SNPs, MAC &gt;= 5, chromosome 1-22, unify SNP names&quot;</span>
    <span class="p">)</span>
    <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">plink2</span><span class="si">}</span><span class="s2"> --pfile </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--allow-extra-chr&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--rm-dup exclude-all&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--max-alleles 2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--mac 5&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--snps-only&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--chr 1-22&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--set-all-var-ids @:#:\$r:\$a&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;--make-pgen --out </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw2&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;--memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">call_helper</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmds</span><span class="p">))</span>

    <span class="c1">## remove SNPs duplicated by position</span>
    <span class="n">df_pvar</span> <span class="o">=</span> <span class="n">dapgen</span><span class="o">.</span><span class="n">read_pvar</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw2.pvar&quot;</span><span class="p">)</span>
    <span class="n">dup_snps</span> <span class="o">=</span> <span class="n">df_pvar</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">df_pvar</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CHROM&quot;</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">])]</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Excluding </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">dup_snps</span><span class="p">)</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">df_pvar</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> SNPs duplicated by positions.&quot;</span>
    <span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw2.snplist&quot;</span><span class="p">,</span> <span class="n">dup_snps</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">plink2</span><span class="si">}</span><span class="s2"> --pfile </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw2&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;--exclude </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw2.snplist&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;--make-pgen --out </span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/all_chr&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;--memory </span><span class="si">{</span><span class="n">memory</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">call_helper</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmds</span><span class="p">))</span>

    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Clean up temporary files...&quot;</span><span class="p">)</span>
    <span class="c1"># remove raw2*</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">dir</span><span class="si">}</span><span class="s2">/pgen/raw2*&quot;</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

    <span class="n">_process_genetic_map</span><span class="p">(</span><span class="n">root_dir</span><span class="o">=</span><span class="nb">dir</span><span class="p">,</span> <span class="n">build</span><span class="o">=</span><span class="n">build</span><span class="p">)</span>
    <span class="n">_process_sample_map</span><span class="p">(</span><span class="n">root_dir</span><span class="o">=</span><span class="nb">dir</span><span class="p">)</span></div>


<div class="viewcode-block" id="select_admix_indiv"><a class="viewcode-back" href="../../../cli/misc.html#admix.cli.select_admix_indiv">[docs]</a><span class="k">def</span> <span class="nf">select_admix_indiv</span><span class="p">(</span>
    <span class="n">ref_pfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">pca_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">superpop1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">superpop2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">out</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">exclude_pop1</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">exclude_pop2</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">n_pc</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>
    <span class="n">sample_t_range</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mf">0.05</span><span class="p">,</span> <span class="mf">0.95</span><span class="p">),</span>
    <span class="n">sample_dist_max</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Select admixed individuals based on joint PCA analysis results of</span>
<span class="sd">    (1) reference dataset (2) sample dataset</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    ref_pfile : str</span>
<span class="sd">        reference panel pfile prefix</span>
<span class="sd">    pca_prefix : str</span>
<span class="sd">        joint pca results prefix. {pca_prefix}.eigenvec, {pca_prefix}.eigenval</span>
<span class="sd">        will be read</span>
<span class="sd">    out : str</span>
<span class="sd">        output prefix</span>
<span class="sd">    superpop1 : str</span>
<span class="sd">        superpopulation 1</span>
<span class="sd">    superpop2 : str</span>
<span class="sd">        superpopulation 2</span>
<span class="sd">    exclude_pop1 : str, optional</span>
<span class="sd">        exclude individuals from superpopulation 1, by default None</span>
<span class="sd">    exclude_pop2 : str, optional</span>
<span class="sd">        exclude individuals from superpopulation 2, by default None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">log_params</span><span class="p">(</span><span class="s2">&quot;select-admixed-indiv&quot;</span><span class="p">,</span> <span class="nb">locals</span><span class="p">())</span>

    <span class="n">df_pc</span><span class="p">,</span> <span class="n">eigenval</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read_joint_pca</span><span class="p">(</span>
        <span class="n">ref_pfile</span><span class="o">=</span><span class="n">ref_pfile</span><span class="p">,</span> <span class="n">pca_prefix</span><span class="o">=</span><span class="n">pca_prefix</span>
    <span class="p">)</span>
    <span class="n">pc_cols</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;PC</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">n_pc</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
    <span class="n">pc_col_pos</span> <span class="o">=</span> <span class="p">[</span><span class="n">df_pc</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">col</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">pc_cols</span><span class="p">]</span>
    <span class="n">pc_eigenval</span> <span class="o">=</span> <span class="n">eigenval</span><span class="p">[</span><span class="n">pc_col_pos</span><span class="p">]</span>

    <span class="n">df_sample_pc</span> <span class="o">=</span> <span class="n">df_pc</span><span class="p">[</span><span class="n">df_pc</span><span class="o">.</span><span class="n">SUPERPOP</span> <span class="o">==</span> <span class="s2">&quot;SAMPLE&quot;</span><span class="p">]</span>
    <span class="n">df_anc1_pc</span> <span class="o">=</span> <span class="n">df_pc</span><span class="p">[</span><span class="n">df_pc</span><span class="o">.</span><span class="n">SUPERPOP</span> <span class="o">==</span> <span class="n">superpop1</span><span class="p">]</span>
    <span class="n">df_anc2_pc</span> <span class="o">=</span> <span class="n">df_pc</span><span class="p">[</span><span class="n">df_pc</span><span class="o">.</span><span class="n">SUPERPOP</span> <span class="o">==</span> <span class="n">superpop2</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">exclude_pop1</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">df_anc1_pc</span> <span class="o">=</span> <span class="n">df_anc1_pc</span><span class="p">[</span><span class="o">~</span><span class="n">df_anc1_pc</span><span class="o">.</span><span class="n">POP</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">exclude_pop1</span><span class="p">)]</span>
    <span class="k">if</span> <span class="n">exclude_pop2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">df_anc2_pc</span> <span class="o">=</span> <span class="n">df_anc2_pc</span><span class="p">[</span><span class="o">~</span><span class="n">df_anc2_pc</span><span class="o">.</span><span class="n">POP</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">exclude_pop2</span><span class="p">)]</span>

    <span class="n">df_sample_pc</span><span class="p">,</span> <span class="n">df_anc1_pc</span><span class="p">,</span> <span class="n">df_anc2_pc</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">df_sample_pc</span><span class="p">[</span><span class="n">pc_cols</span><span class="p">],</span>
        <span class="n">df_anc1_pc</span><span class="p">[</span><span class="n">pc_cols</span><span class="p">],</span>
        <span class="n">df_anc2_pc</span><span class="p">[</span><span class="n">pc_cols</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="c1"># prepare sample_pc, anc1_pc, anc2_pc</span>
    <span class="n">sample_dist</span><span class="p">,</span> <span class="n">sample_t</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">distance_to_refpop</span><span class="p">(</span>
        <span class="n">df_sample_pc</span><span class="p">,</span> <span class="n">df_anc1_pc</span><span class="p">,</span> <span class="n">df_anc2_pc</span><span class="p">,</span> <span class="n">weight</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">pc_eigenval</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="n">selected_mask</span> <span class="o">=</span> <span class="p">(</span>
        <span class="p">(</span><span class="n">sample_t_range</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">sample_t</span><span class="p">)</span>
        <span class="o">&amp;</span> <span class="p">(</span><span class="n">sample_t</span> <span class="o">&lt;</span> <span class="n">sample_t_range</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="o">&amp;</span> <span class="p">(</span><span class="n">sample_dist</span> <span class="o">&lt;</span> <span class="n">sample_dist_max</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">selected_indiv</span> <span class="o">=</span> <span class="n">df_sample_pc</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">selected_mask</span><span class="p">]</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">selected_indiv</span><span class="p">)</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">df_sample_pc</span><span class="p">)</span><span class="si">}</span><span class="s2"> selected to be admixed individuals.&quot;</span>
    <span class="p">)</span>

    <span class="n">df_plot</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df_pc</span><span class="p">[</span><span class="n">df_pc</span><span class="o">.</span><span class="n">SUPERPOP</span> <span class="o">!=</span> <span class="s2">&quot;SAMPLE&quot;</span><span class="p">],</span> <span class="n">df_pc</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">selected_indiv</span><span class="p">]])</span>
    <span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mf">8.5</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">joint_pca</span><span class="p">(</span><span class="n">df_pc</span><span class="o">=</span><span class="n">df_plot</span><span class="p">,</span> <span class="n">eigenval</span><span class="o">=</span><span class="n">eigenval</span><span class="p">,</span> <span class="n">axes</span><span class="o">=</span><span class="n">axes</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out</span><span class="si">}</span><span class="s2">.indiv&quot;</span><span class="p">,</span> <span class="n">selected_indiv</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PCA plots saved to </span><span class="si">{</span><span class="n">out</span><span class="si">}</span><span class="s2">.png.&quot;</span><span class="p">)</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Selected individuals saved to </span><span class="si">{</span><span class="n">out</span><span class="si">}</span><span class="s2">.indiv.&quot;</span><span class="p">)</span></div>
</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021-2024, Kangcheng Hou
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/kangchenghou/admix-kit" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js?v=b3ba4146"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../../../_static/scripts/furo.js?v=32e29ea5"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/tabs.js?v=3ee01567"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>