<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2024.01.29 -->
        <title>admix.tools.plink2 - admix-kit docs</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/tabs.css?v=4c969af8" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">admix-kit docs</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  
  
  <span class="sidebar-brand-text">admix-kit docs</span>
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../prepare-dataset.html">Dataset preparation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../cli/index.html">Command line interface</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Command line interface</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/assoc-test.html">Association testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/genet-cor.html">Genetic correlation estimation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/pgs.html">Local ancestry-specific polygenic score</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/simulate-pheno.html">Phenotype simulation</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../api.html">Python API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Python API</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.assoc.marginal.html">admix.assoc.marginal</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.admix_grm.html">admix.cli.admix_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.admix_grm_merge.html">admix.cli.admix_grm_merge</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.genet_cor.html">admix.cli.genet_cor</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.calc_partial_pgs.html">admix.data.calc_partial_pgs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.Dataset.html">admix.Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.admix_geno.html">admix.simulate.admix_geno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno.html">admix.simulate.quant_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.binary_pheno.html">admix.simulate.binary_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno_1pop.html">admix.simulate.quant_pheno_1pop</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno_grm.html">admix.simulate.quant_pheno_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.allele_per_anc.html">admix.data.allele_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.af_per_anc.html">admix.data.af_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_dataset.html">admix.io.read_dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_lanc.html">admix.io.read_lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_rfmix.html">admix.io.read_rfmix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.tools.plink2.lift_over.html">admix.tools.plink2.lift_over</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.lanc.html">admix.plot.lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.admixture.html">admix.plot.admixture</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.manhattan.html">admix.plot.manhattan</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.qq.html">admix.plot.qq</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Example usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../simulate-admix-genotype.html">Admixed genotype simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/analyze-admix-simu-data.html">Basic statistics and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/assoc.html">Association testing (GWAS)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/genet-cor.html">Genetic correlation across local ancestry segments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/polygenic-scoring.html">Polygenic scoring for each local ancestry background</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/dataset.html">admix.Dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/select-admixed.html">Selecting admixed individuals</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <h1>Source code for admix.tools.plink2</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">dapgen</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">admix</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="kn">import</span> <span class="nn">dapgen</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">admix</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">from</span> <span class="nn">._utils</span> <span class="kn">import</span> <span class="n">get_dependency</span>


<span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Shortcut for running plink commands</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    cmd : str</span>
<span class="sd">        plink command</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bin_path</span> <span class="o">=</span> <span class="n">get_dependency</span><span class="p">(</span><span class="s2">&quot;plink2&quot;</span><span class="p">)</span>
    <span class="n">add_cmds</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot; --</span><span class="si">{</span><span class="n">k</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;-&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">kwargs</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="p">]</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">add_cmds</span><span class="p">)</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">check_call</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bin_path</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">cmd</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">gwas</span><span class="p">(</span>
    <span class="n">df_sample_info</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">pheno_col</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">out_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">pfile</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">bfile</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">family</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;linear&quot;</span><span class="p">,</span>
    <span class="n">covar_cols</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">cat_cols</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">pheno_quantile_normalize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">covar_quantile_normalize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">clean_tmp_file</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="p">):</span>
    <span class="k">assert</span> <span class="n">family</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;linear&quot;</span><span class="p">,</span> <span class="s2">&quot;logistic&quot;</span><span class="p">],</span> <span class="s2">&quot;family must be linear or logistic&quot;</span>
    <span class="c1"># only one of pfile or bfile must be provided</span>
    <span class="k">assert</span> <span class="p">(</span><span class="n">pfile</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="o">!=</span> <span class="p">(</span><span class="n">bfile</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">),</span> <span class="s2">&quot;pfile or bfile must be provided&quot;</span>

    <span class="c1"># clean up phenotype</span>
    <span class="n">non_nan_index</span> <span class="o">=</span> <span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">df_sample_info</span><span class="p">[</span><span class="n">pheno_col</span><span class="p">])</span>
    <span class="n">df_pheno</span> <span class="o">=</span> <span class="n">df_sample_info</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">non_nan_index</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="n">pheno_col</span><span class="p">:</span> <span class="s2">&quot;trait&quot;</span><span class="p">})</span>
    <span class="n">pheno_vals</span> <span class="o">=</span> <span class="n">df_pheno</span><span class="o">.</span><span class="n">trait</span><span class="o">.</span><span class="n">values</span>
    <span class="k">if</span> <span class="n">family</span> <span class="o">==</span> <span class="s2">&quot;logistic&quot;</span><span class="p">:</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">pheno_quantile_normalize</span> <span class="o">==</span> <span class="kc">False</span>
        <span class="p">),</span> <span class="s2">&quot;quantile normalization can not be used for logistic regression&quot;</span>
        <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
            <span class="p">(</span><span class="n">pheno_vals</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">pheno_vals</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
        <span class="p">),</span> <span class="s2">&quot;phenotype values must be 0 or 1 for logistic regression&quot;</span>
        <span class="n">df_pheno</span><span class="p">[</span><span class="s2">&quot;trait&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_pheno</span><span class="o">.</span><span class="n">trait</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="k">elif</span> <span class="n">family</span> <span class="o">==</span> <span class="s2">&quot;linear&quot;</span><span class="p">:</span>
        <span class="n">df_pheno</span><span class="p">[</span><span class="s2">&quot;trait&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">pheno_vals</span> <span class="o">-</span> <span class="n">pheno_vals</span><span class="o">.</span><span class="n">mean</span><span class="p">())</span> <span class="o">/</span> <span class="n">pheno_vals</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;family must be linear or logistic&quot;</span><span class="p">)</span>
    <span class="n">pheno_path</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;.plink2_tmp_pheno&quot;</span>

    <span class="k">if</span> <span class="n">bfile</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># PLINK1 format</span>
        <span class="c1"># FID, IID must be in the column</span>
        <span class="k">assert</span> <span class="s2">&quot;FID&quot;</span> <span class="ow">in</span> <span class="n">df_sample_info</span><span class="o">.</span><span class="n">columns</span>
        <span class="k">assert</span> <span class="s2">&quot;IID&quot;</span> <span class="ow">in</span> <span class="n">df_sample_info</span><span class="o">.</span><span class="n">columns</span>
        <span class="n">df_pheno</span><span class="p">[[</span><span class="s2">&quot;FID&quot;</span><span class="p">,</span> <span class="s2">&quot;IID&quot;</span><span class="p">,</span> <span class="s2">&quot;trait&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
            <span class="n">pheno_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">na_rep</span><span class="o">=</span><span class="s2">&quot;NA&quot;</span>
        <span class="p">)</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;--bfile </span><span class="si">{</span><span class="n">bfile</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">pfile</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># PLINK2 format</span>
        <span class="n">df_pheno</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;#IID&quot;</span>
        <span class="n">df_pheno</span><span class="p">[[</span><span class="s2">&quot;trait&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">pheno_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">na_rep</span><span class="o">=</span><span class="s2">&quot;NA&quot;</span><span class="p">)</span>
        <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;--pfile </span><span class="si">{</span><span class="n">pfile</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>

    <span class="n">cmds</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="sa">f</span><span class="s2">&quot;--pheno </span><span class="si">{</span><span class="n">pheno_path</span><span class="si">}</span><span class="s2"> --no-psam-pheno&quot;</span><span class="p">,</span>
            <span class="sa">f</span><span class="s2">&quot;--out </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="n">family</span> <span class="o">==</span> <span class="s2">&quot;linear&quot;</span><span class="p">:</span>
        <span class="n">cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--linear hide-covar omit-ref&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">family</span> <span class="o">==</span> <span class="s2">&quot;logistic&quot;</span><span class="p">:</span>
        <span class="n">cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--logistic hide-covar omit-ref no-firth&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;family must be linear or logistic&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">covar_cols</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">covar_path</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="sa">f</span><span class="s2">&quot;.plink2_tmp_covar&quot;</span>
        <span class="n">cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--covar </span><span class="si">{</span><span class="n">covar_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">bfile</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># PLINK1 format</span>
            <span class="c1"># FID, IID must be in the column</span>
            <span class="k">assert</span> <span class="s2">&quot;FID&quot;</span> <span class="ow">in</span> <span class="n">df_sample_info</span><span class="o">.</span><span class="n">columns</span>
            <span class="k">assert</span> <span class="s2">&quot;IID&quot;</span> <span class="ow">in</span> <span class="n">df_sample_info</span><span class="o">.</span><span class="n">columns</span>
            <span class="n">df_covar</span> <span class="o">=</span> <span class="n">df_sample_info</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                <span class="n">non_nan_index</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;FID&quot;</span><span class="p">,</span> <span class="s2">&quot;IID&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">covar_cols</span>
            <span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">df_covar</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">covar_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">na_rep</span><span class="o">=</span><span class="s2">&quot;NA&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">pfile</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># PLINK2 format</span>
            <span class="n">df_covar</span> <span class="o">=</span> <span class="n">df_sample_info</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">non_nan_index</span><span class="p">,</span> <span class="n">covar_cols</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">df_covar</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;#IID&quot;</span>
            <span class="n">df_covar</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">covar_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">na_rep</span><span class="o">=</span><span class="s2">&quot;NA&quot;</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">cmds</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="s2">&quot; allow-no-covars&quot;</span>

    <span class="k">if</span> <span class="n">pheno_quantile_normalize</span><span class="p">:</span>
        <span class="n">cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--pheno-quantile-normalize&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">covar_quantile_normalize</span><span class="p">:</span>
        <span class="n">cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--covar-quantile-normalize&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">cat_cols</span><span class="p">:</span>
        <span class="n">cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--split-cat-pheno omit-most </span><span class="si">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cat_cols</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmds</span><span class="p">))</span>

    <span class="n">run</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmds</span><span class="p">),</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">family</span> <span class="o">==</span> <span class="s2">&quot;linear&quot;</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.trait.glm.linear&quot;</span><span class="p">,</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.assoc&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.trait.glm.logistic&quot;</span><span class="p">,</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.assoc&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">clean_tmp_file</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.plink2_tmp_*&quot;</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">clump</span><span class="p">(</span>
    <span class="n">pfile</span><span class="p">,</span>
    <span class="n">assoc_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">out_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">p1</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5e-8</span><span class="p">,</span>
    <span class="n">p2</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1e-4</span><span class="p">,</span>
    <span class="n">r2</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span>
    <span class="n">kb</span><span class="o">=</span><span class="mi">3000</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Wrapper for plink2 clump</span>
<span class="sd">    For now, first need to export to .bed format then perform the clump</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tmp_prefix</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.admix_plink2_clump_tmp&quot;</span>
    <span class="c1"># convert to bed</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">plink2</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--pfile </span><span class="si">{</span><span class="n">pfile</span><span class="si">}</span><span class="s2"> --make-bed --out </span><span class="si">{</span><span class="n">tmp_prefix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># perform clump</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">plink</span><span class="o">.</span><span class="n">clump</span><span class="p">(</span>
        <span class="n">bfile</span><span class="o">=</span><span class="n">tmp_prefix</span><span class="p">,</span>
        <span class="n">assoc_path</span><span class="o">=</span><span class="n">assoc_path</span><span class="p">,</span>
        <span class="n">out_prefix</span><span class="o">=</span><span class="n">out_prefix</span><span class="p">,</span>
        <span class="n">p1</span><span class="o">=</span><span class="n">p1</span><span class="p">,</span>
        <span class="n">p2</span><span class="o">=</span><span class="n">p2</span><span class="p">,</span>
        <span class="n">r2</span><span class="o">=</span><span class="n">r2</span><span class="p">,</span>
        <span class="n">kb</span><span class="o">=</span><span class="n">kb</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># # convert plink2 association to plink1 format ID -&gt; SNP</span>
    <span class="c1"># import shutil</span>

    <span class="c1"># from_file = open(assoc_path)</span>
    <span class="c1"># to_file = open(tmp_prefix + &quot;.assoc&quot;, &quot;w&quot;)</span>
    <span class="c1"># to_file.writelines(from_file.readline().replace(&quot;ID&quot;, &quot;SNP&quot;))</span>
    <span class="c1"># shutil.copyfileobj(from_file, to_file)</span>
    <span class="c1"># from_file.close()</span>
    <span class="c1"># to_file.close()</span>
    <span class="c1"># cmds = [</span>
    <span class="c1">#     f&quot;--bfile {tmp_prefix} --clump {tmp_prefix + &#39;.assoc&#39;}&quot;,</span>
    <span class="c1">#     f&quot;--clump-p1 {p1} --clump-p2 {p2} --clump-r2 {r2} --clump-kb {kb}&quot;,</span>
    <span class="c1">#     f&quot;--out {tmp_prefix}&quot;,</span>
    <span class="c1"># ]</span>

    <span class="c1"># admix.tools.plink.run(&quot; &quot;.join(cmds))</span>
    <span class="c1"># if os.path.exists(tmp_prefix + &quot;.clumped&quot;):</span>
    <span class="c1">#     os.rename(tmp_prefix + &quot;.clumped&quot;, out_prefix + &quot;.clumped&quot;)</span>
    <span class="c1"># else:</span>
    <span class="c1">#     # no clumped region</span>
    <span class="c1">#     # write a comment to the output file</span>
    <span class="c1">#     with open(out_prefix + &quot;.clumped&quot;, &quot;w&quot;) as file:</span>
    <span class="c1">#         file.write(&quot;# No clumped region&quot;)</span>

    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">tmp_prefix</span> <span class="o">+</span> <span class="s2">&quot;*&quot;</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>


<div class="viewcode-block" id="lift_over"><a class="viewcode-back" href="../../../_api/admix.tools.plink2.lift_over.html#admix.tools.plink2.lift_over">[docs]</a><span class="k">def</span> <span class="nf">lift_over</span><span class="p">(</span><span class="n">pfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="s2">&quot;hg19-&gt;hg38&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Lift over a plink file to another genome. First, use liftover to convert to</span>
<span class="sd">    another genome, then use plink2 --sort-vars to sort the variants.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    pfile : str</span>
<span class="sd">        Path to plink file</span>
<span class="sd">    out_prefix : str</span>
<span class="sd">        output prefix, &lt;out_prefix&gt;.pgen, &lt;out_prefix&gt;.psam, &lt;out_prefix&gt;.pvar will</span>
<span class="sd">        be generated</span>
<span class="sd">    chain : str, optional</span>
<span class="sd">        lift over direction, by default &quot;hg19-&gt;hg38&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">chain</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;hg19-&gt;hg38&quot;</span><span class="p">,</span> <span class="s2">&quot;hg38-&gt;hg19&quot;</span><span class="p">]</span>

    <span class="c1"># read the input pgen</span>
    <span class="n">pgen</span><span class="p">,</span> <span class="n">pvar</span><span class="p">,</span> <span class="n">psam</span> <span class="o">=</span> <span class="n">dapgen</span><span class="o">.</span><span class="n">read_pfile</span><span class="p">(</span><span class="n">pfile</span><span class="p">)</span>
    <span class="n">df_snp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="p">{</span><span class="s2">&quot;CHROM&quot;</span><span class="p">:</span> <span class="n">pvar</span><span class="o">.</span><span class="n">CHROM</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">:</span> <span class="n">pvar</span><span class="o">.</span><span class="n">POS</span><span class="o">.</span><span class="n">values</span><span class="p">},</span> <span class="n">index</span><span class="o">=</span><span class="n">pvar</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>
    <span class="p">)</span>
    <span class="c1"># perform the liftover</span>
    <span class="n">df_lifted</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">liftover</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">df_snp</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="n">chain</span><span class="p">)</span>
    <span class="n">n_snp1</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_lifted</span><span class="p">)</span>
    <span class="n">df_lifted</span> <span class="o">=</span> <span class="n">df_lifted</span><span class="p">[</span><span class="n">df_lifted</span><span class="o">.</span><span class="n">POS</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">n_snp2</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_lifted</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;remove </span><span class="si">{</span><span class="n">n_snp1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">n_snp2</span><span class="si">}</span><span class="s2"> / </span><span class="si">{</span><span class="n">n_snp1</span><span class="si">}</span><span class="s2"> SNPs&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; (</span><span class="si">{</span><span class="p">(</span><span class="n">n_snp1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">n_snp2</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">n_snp1</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="si">:</span><span class="s2">.2g</span><span class="si">}</span><span class="s2">%)&quot;</span>
        <span class="s2">&quot; for unmapped or ambiguous SNPs&quot;</span>
    <span class="p">)</span>

    <span class="c1"># extract the lifted SNPs</span>
    <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">-tmp.snp&quot;</span><span class="p">,</span> <span class="n">df_lifted</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">plink2</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;--pfile </span><span class="si">{</span><span class="n">pfile</span><span class="si">}</span><span class="s2"> --extract </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">-tmp.snp --sort-vars --make-pgen --out </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">-tmp&quot;</span>
    <span class="p">)</span>

    <span class="c1"># substitute the coordinates obtained from liftover</span>
    <span class="n">pgen</span><span class="p">,</span> <span class="n">pvar</span><span class="p">,</span> <span class="n">psam</span> <span class="o">=</span> <span class="n">dapgen</span><span class="o">.</span><span class="n">read_pfile</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;-tmp&quot;</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">pvar</span><span class="o">.</span><span class="n">index</span> <span class="o">==</span> <span class="n">df_lifted</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

    <span class="n">pvar</span><span class="p">[</span><span class="s2">&quot;POS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_lifted</span><span class="p">[</span><span class="s2">&quot;POS&quot;</span><span class="p">]</span>
    <span class="n">pvar</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;ID&quot;</span><span class="p">,</span> <span class="n">pvar</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="n">pvar</span> <span class="o">=</span> <span class="n">pvar</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;CHROM&quot;</span><span class="p">:</span> <span class="s2">&quot;#CHROM&quot;</span><span class="p">})</span>
    <span class="n">pvar</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">-tmp.pvar&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">plink2</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;--pfile </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">-tmp --sort-vars --make-pgen --out </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>
    <span class="c1"># remove the tmp files</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;-tmp*&quot;</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">f</span><span class="p">)</span></div>


<span class="k">def</span> <span class="nf">align_snp</span><span class="p">(</span><span class="n">pfile1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">pfile2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Given two plink files from different individuals, take the common set of SNPs.</span>
<span class="sd">    Matching by (1) chromosome (2) position (3) ref allele (4) alt allele</span>

<span class="sd">    This can be useful for software that requires an aligned set of SNPs</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    pfile1 : str</span>
<span class="sd">        plink file 1</span>
<span class="sd">    pfile2 : str</span>
<span class="sd">        plink file 2</span>
<span class="sd">    out_prefix: str</span>
<span class="sd">        prefix to the output files</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Step 1: read pvar and find the common SNPs</span>
    <span class="n">df_snp1</span> <span class="o">=</span> <span class="n">dapgen</span><span class="o">.</span><span class="n">read_pvar</span><span class="p">(</span><span class="n">pfile1</span> <span class="o">+</span> <span class="s2">&quot;.pvar&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span>
        <span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CHROM&quot;</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">df_snp2</span> <span class="o">=</span> <span class="n">dapgen</span><span class="o">.</span><span class="n">read_pvar</span><span class="p">(</span><span class="n">pfile2</span> <span class="o">+</span> <span class="s2">&quot;.pvar&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span>
        <span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CHROM&quot;</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">]</span>
    <span class="p">)</span>

    <span class="n">idx1</span><span class="p">,</span> <span class="n">idx2</span><span class="p">,</span> <span class="n">flip</span> <span class="o">=</span> <span class="n">dapgen</span><span class="o">.</span><span class="n">align_snp</span><span class="p">(</span><span class="n">df_snp1</span><span class="p">,</span> <span class="n">df_snp2</span><span class="p">)</span>

    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">idx1</span><span class="p">)</span><span class="si">}</span><span class="s2"> / </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">df_snp1</span><span class="p">)</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">idx1</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">df_snp1</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="si">:</span><span class="s2">.3g</span><span class="si">}</span><span class="s2">%) SNPs extracted in pfile1: </span><span class="si">{</span><span class="n">pfile1</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">idx2</span><span class="p">)</span><span class="si">}</span><span class="s2"> / </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">df_snp2</span><span class="p">)</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">idx2</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nb">len</span><span class="p">(</span><span class="n">df_snp2</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="mi">100</span><span class="si">:</span><span class="s2">.3g</span><span class="si">}</span><span class="s2">%) SNPs extracted in pfile2: </span><span class="si">{</span><span class="n">pfile2</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Step 2: extract the common SNPs from the two files and fix the allele order</span>
    <span class="n">allele1</span> <span class="o">=</span> <span class="n">df_snp1</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">idx1</span><span class="p">][</span><span class="s2">&quot;REF&quot;</span><span class="p">]</span>
    <span class="n">allele2</span> <span class="o">=</span> <span class="n">df_snp2</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">idx2</span><span class="p">][</span><span class="s2">&quot;REF&quot;</span><span class="p">]</span>

    <span class="n">allele1</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;-tmp.1.refallele&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;-tmp.1.snplist&quot;</span><span class="p">,</span> <span class="n">allele1</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># use sample ref-allele values, but use ref-allele index</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">allele1</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">allele2</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
        <span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;-tmp.2.refallele&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;-tmp.2.snplist&quot;</span><span class="p">,</span> <span class="n">allele2</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># extract common SNPs and align REF ALT</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">plink2</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;--pfile </span><span class="si">{</span><span class="n">pfile1</span><span class="si">}</span><span class="s2"> --extract </span><span class="si">{</span><span class="n">out_prefix</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s1">&#39;-tmp.1.snplist&#39;</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --ref-allele </span><span class="si">{</span><span class="n">out_prefix</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s1">&#39;-tmp.1.refallele&#39;</span><span class="si">}</span><span class="s2"> 2 1 --make-pgen --out </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">.1&quot;</span>
    <span class="p">)</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">plink2</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;--pfile </span><span class="si">{</span><span class="n">pfile2</span><span class="si">}</span><span class="s2"> --extract </span><span class="si">{</span><span class="n">out_prefix</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s1">&#39;-tmp.2.snplist&#39;</span><span class="si">}</span><span class="s2"> &quot;</span>
        <span class="sa">f</span><span class="s2">&quot; --ref-allele force </span><span class="si">{</span><span class="n">out_prefix</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s1">&#39;-tmp.2.refallele&#39;</span><span class="si">}</span><span class="s2"> 2 1 --make-pgen --out </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">.2&quot;</span>
    <span class="p">)</span>

    <span class="c1"># remove the tmp files</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;-tmp*&quot;</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">merge_indiv</span><span class="p">(</span><span class="n">pfile1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">pfile2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Given two plink files from different individuals, take the common set of SNPs</span>
<span class="sd">    and merge them into one data set.</span>

<span class="sd">    This is a wrapper for various PLINK functions. Since PLINK2&#39;s pmerge function is</span>
<span class="sd">    not complete now. We use PLINK1 merge functionality instead. A PLINK1 bfile is</span>
<span class="sd">    produced (rather than PLINK2 pfile).</span>

<span class="sd">    This can be useful for example when we want to perform a joint PCA for the two</span>
<span class="sd">    data sets.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    pfile1 : str</span>
<span class="sd">        plink file 1</span>
<span class="sd">    pfile2 : str</span>
<span class="sd">        plink file 2</span>
<span class="sd">    out_prefix: str</span>
<span class="sd">        prefix to the output files</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">plink2</span><span class="o">.</span><span class="n">align_snp</span><span class="p">(</span>
        <span class="n">pfile1</span><span class="o">=</span><span class="n">pfile1</span><span class="p">,</span> <span class="n">pfile2</span><span class="o">=</span><span class="n">pfile2</span><span class="p">,</span> <span class="n">out_prefix</span><span class="o">=</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;-tmp&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Step 3: unify SNP ID</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">plink2</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;--pfile </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">-tmp.1 --set-all-var-ids @:#:\$r:\$a --make-bed --out </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">-tmp.1.fixid&quot;</span>
    <span class="p">)</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">plink2</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;--pfile </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">-tmp.2 --set-all-var-ids @:#:\$r:\$a --make-bed --out </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">-tmp.2.fixid&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Step 4: make final merged bfile</span>
    <span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">plink</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;--bfile </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">-tmp.1.fixid --bmerge </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">-tmp.2.fixid --make-bed --out </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>

    <span class="c1"># remove the tmp files</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;-tmp*&quot;</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">subset</span><span class="p">(</span>
    <span class="n">pfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">out_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">snp_list</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">indiv_list</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
    <span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;--pfile </span><span class="si">{</span><span class="n">pfile</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;--make-pgen --out </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;--memory 16000&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">snp_list</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">snplist_path</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.admix-plink2-subset.snplist&quot;</span>
        <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="n">snplist_path</span><span class="p">,</span> <span class="n">snp_list</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--extract </span><span class="si">{</span><span class="n">snplist_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">indiv_list</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">indivlist_path</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.admix-plink2-subset.indivlist&quot;</span>
        <span class="n">np</span><span class="o">.</span><span class="n">savetxt</span><span class="p">(</span><span class="n">indivlist_path</span><span class="p">,</span> <span class="n">indiv_list</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--keep </span><span class="si">{</span><span class="n">indivlist_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">run</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmds</span><span class="p">))</span>

    <span class="c1"># clean up</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.admix-plink2-subset.*&quot;</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>


<span class="c1"># TODO: check difference with plink2.gwas, consider merging these two functions</span>
<span class="k">def</span> <span class="nf">assoc</span><span class="p">(</span>
    <span class="n">bfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">pheno</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">out_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">covar</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">indiv</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">snp</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Run plink association test</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bfile : str</span>
<span class="sd">        plink binary file</span>
<span class="sd">    pheno : pd.DataFrame</span>
<span class="sd">        phenotype data</span>
<span class="sd">    out_prefix : str</span>
<span class="sd">        prefix for output files</span>
<span class="sd">    covar : pd.DataFrame, optional</span>
<span class="sd">        covariate data</span>
<span class="sd">    indiv : pd.DataFrame, optional</span>
<span class="sd">        individual to subset, should contain columns FID and IID</span>
<span class="sd">    snp : pd.DataFrame, optional</span>
<span class="sd">        snp to subset, should contain column SNP</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">assert</span> <span class="n">indiv</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">snp</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;indiv and snp are not supported&quot;</span>
    <span class="n">pheno_path</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.tmp_pheno&quot;</span>
    <span class="n">covar_path</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.tmp_covar&quot;</span>
    <span class="n">pheno</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">pheno_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">covar</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">covar</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">covar_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span>
        <span class="sa">f</span><span class="s2">&quot;--bfile </span><span class="si">{</span><span class="n">bfile</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;--out </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;--pheno </span><span class="si">{</span><span class="n">pheno_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="k">if</span> <span class="n">covar</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">cmd</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--ci 0.95 --glm omit-ref hide-covar&quot;</span><span class="p">)</span>
        <span class="n">cmd</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--covar </span><span class="si">{</span><span class="n">covar_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">cmd</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--ci 0.95 --glm omit-ref hide-covar allow-no-covars&quot;</span><span class="p">)</span>

    <span class="n">run</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmd</span><span class="p">))</span>


<span class="k">def</span> <span class="nf">prune</span><span class="p">(</span>
    <span class="n">pfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">out_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">indep_params</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">indep_pairwise_params</span><span class="p">:</span> <span class="n">List</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Run plink2 prune</span>

<span class="sd">    For example, indep_params = [200, 5, 1.15], indep-pairwise = [100 5 0.1], then two-step</span>
<span class="sd">    pruning will be performed.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    pfile : str</span>
<span class="sd">        plink binary file</span>
<span class="sd">    indep_params : List</span>
<span class="sd">        list of parameters for indep</span>
<span class="sd">    indep_pairwise_params : List</span>
<span class="sd">        list of parameters for indep-pairwise</span>
<span class="sd">    out_prefix : str</span>
<span class="sd">        prefix for output files</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">assert</span> <span class="n">indep_params</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;indep_params is not supported yet&quot;</span>

    <span class="n">tmp_prefix</span> <span class="o">=</span> <span class="n">out_prefix</span> <span class="o">+</span> <span class="s2">&quot;.admix_plink2_prune_tmp&quot;</span>

    <span class="k">assert</span> <span class="p">(</span><span class="n">indep_params</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="o">!=</span> <span class="p">(</span>
        <span class="n">indep_pairwise_params</span> <span class="ow">is</span> <span class="kc">None</span>
    <span class="p">),</span> <span class="s2">&quot;only one of indep_params and indep_pairwise_params can be specified&quot;</span>

    <span class="c1"># step 1: indep pruning</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span>
        <span class="sa">f</span><span class="s2">&quot;--pfile </span><span class="si">{</span><span class="n">pfile</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;--out </span><span class="si">{</span><span class="n">tmp_prefix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">if</span> <span class="n">indep_params</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">cmd</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;--indep </span><span class="si">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">p</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">p</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">indep_params</span><span class="p">])</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">indep_pairwise_params</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">cmd</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;--indep-pairwise </span><span class="si">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">p</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">p</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">indep_pairwise_params</span><span class="p">])</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="n">run</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmd</span><span class="p">))</span>

    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span>
        <span class="sa">f</span><span class="s2">&quot;--pfile </span><span class="si">{</span><span class="n">pfile</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;--extract </span><span class="si">{</span><span class="n">tmp_prefix</span><span class="si">}</span><span class="s2">.prune.in&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;--make-pgen --out </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">run</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmd</span><span class="p">))</span>

    <span class="c1"># clean up</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">tmp_prefix</span> <span class="o">+</span> <span class="s2">&quot;.*&quot;</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">pca</span><span class="p">(</span><span class="n">pfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">approx</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Run plink2 pca</span>

<span class="sd">    TODO: include more options</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    pfile : str</span>
<span class="sd">        plink binary file</span>
<span class="sd">    out_prefix : str</span>
<span class="sd">        prefix for output files</span>
<span class="sd">    approx: bool</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span>
        <span class="sa">f</span><span class="s2">&quot;--pfile </span><span class="si">{</span><span class="n">pfile</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;--out </span><span class="si">{</span><span class="n">out_prefix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="k">if</span> <span class="n">approx</span><span class="p">:</span>
        <span class="n">cmd</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--pca approx&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">cmd</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--pca&quot;</span><span class="p">)</span>
    <span class="n">run</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cmd</span><span class="p">),</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">subset_hapmap3</span><span class="p">(</span><span class="n">pfile</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">build</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">chrom</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Overlap a pfile with HapMap3 SNPs</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    pfile : str</span>
<span class="sd">        input pfile</span>
<span class="sd">    build: str</span>
<span class="sd">        hg19 or hg38</span>
<span class="sd">    out_prefix: str</span>
<span class="sd">        output pfile prefix</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    &lt;out_prefix&gt;.{pgen|pvar|psam} will be written to disk if out_prefix is specified</span>
<span class="sd">    otherwise, snp_list will be returned</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">build</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">],</span> <span class="s2">&quot;build must be hg19 or hg38&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">pyreadr</span>
    <span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ImportError</span><span class="p">(</span><span class="s2">&quot;pyreadr is not installed. Use pip install pyreadr&quot;</span><span class="p">)</span>
    <span class="n">hm3_snps</span> <span class="o">=</span> <span class="n">pyreadr</span><span class="o">.</span><span class="n">read_r</span><span class="p">(</span><span class="n">admix</span><span class="o">.</span><span class="n">tools</span><span class="o">.</span><span class="n">get_cache_data</span><span class="p">(</span><span class="s2">&quot;hapmap3_snps&quot;</span><span class="p">))[</span><span class="kc">None</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">build</span> <span class="o">==</span> <span class="s2">&quot;hg19&quot;</span><span class="p">:</span>
        <span class="n">hm3_snps</span> <span class="o">=</span> <span class="n">hm3_snps</span><span class="p">[[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;chr&quot;</span><span class="p">:</span> <span class="s2">&quot;CHROM&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">:</span> <span class="s2">&quot;POS&quot;</span><span class="p">}</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">build</span> <span class="o">==</span> <span class="s2">&quot;hg38&quot;</span><span class="p">:</span>
        <span class="n">hm3_snps</span> <span class="o">=</span> <span class="n">hm3_snps</span><span class="p">[[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;pos_hg38&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;chr&quot;</span><span class="p">:</span> <span class="s2">&quot;CHROM&quot;</span><span class="p">,</span> <span class="s2">&quot;pos_hg38&quot;</span><span class="p">:</span> <span class="s2">&quot;POS&quot;</span><span class="p">}</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span>

    <span class="n">TOTAL_HM3_SNPS</span> <span class="o">=</span> <span class="n">hm3_snps</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">chrom</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">hm3_snps</span> <span class="o">=</span> <span class="n">hm3_snps</span><span class="p">[</span><span class="n">hm3_snps</span><span class="o">.</span><span class="n">CHROM</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">]</span>

    <span class="n">df_snp</span> <span class="o">=</span> <span class="n">dapgen</span><span class="o">.</span><span class="n">read_pvar</span><span class="p">(</span><span class="n">pfile</span> <span class="o">+</span> <span class="s2">&quot;.pvar&quot;</span><span class="p">)</span>
    <span class="n">snp_list</span> <span class="o">=</span> <span class="n">df_snp</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">hm3_snps</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CHROM&quot;</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">])[</span><span class="s2">&quot;snp&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

    <span class="n">admix</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">snp_list</span><span class="p">)</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">TOTAL_HM3_SNPS</span><span class="si">}</span><span class="s2"> SNPs are retained&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">out_prefix</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">snp_list</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">subset</span><span class="p">(</span><span class="n">pfile</span><span class="p">,</span> <span class="n">out_prefix</span><span class="p">,</span> <span class="n">snp_list</span><span class="o">=</span><span class="n">snp_list</span><span class="p">)</span>
</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021-2024, Kangcheng Hou
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/kangchenghou/admix-kit" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js?v=b3ba4146"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../../../_static/scripts/furo.js?v=32e29ea5"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/tabs.js?v=3ee01567"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>