<!doctype html>
<html class="no-js" lang="en" data-content_root="">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" />

    <!-- Generated with Sphinx 7.1.2 and Furo 2024.01.29 -->
        <title>admix.simulate._geno - admix-kit docs</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/tabs.css?v=4c969af8" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">admix-kit docs</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  
  
  <span class="sidebar-brand-text">admix-kit docs</span>
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../prepare-dataset.html">Dataset preparation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../cli/index.html">Command line interface</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Command line interface</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/assoc-test.html">Association testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/genet-cor.html">Genetic correlation estimation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/pgs.html">Local ancestry-specific polygenic score</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cli/simulate-pheno.html">Phenotype simulation</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../api.html">Python API</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Python API</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.assoc.marginal.html">admix.assoc.marginal</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.admix_grm.html">admix.cli.admix_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.admix_grm_merge.html">admix.cli.admix_grm_merge</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.cli.genet_cor.html">admix.cli.genet_cor</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.calc_partial_pgs.html">admix.data.calc_partial_pgs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.Dataset.html">admix.Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.admix_geno.html">admix.simulate.admix_geno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno.html">admix.simulate.quant_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.binary_pheno.html">admix.simulate.binary_pheno</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno_1pop.html">admix.simulate.quant_pheno_1pop</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.simulate.quant_pheno_grm.html">admix.simulate.quant_pheno_grm</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.allele_per_anc.html">admix.data.allele_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.data.af_per_anc.html">admix.data.af_per_anc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_dataset.html">admix.io.read_dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_lanc.html">admix.io.read_lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.io.read_rfmix.html">admix.io.read_rfmix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.tools.plink2.lift_over.html">admix.tools.plink2.lift_over</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.lanc.html">admix.plot.lanc</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.admixture.html">admix.plot.admixture</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.manhattan.html">admix.plot.manhattan</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../_api/admix.plot.qq.html">admix.plot.qq</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Example usage</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../simulate-admix-genotype.html">Admixed genotype simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/analyze-admix-simu-data.html">Basic statistics and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/assoc.html">Association testing (GWAS)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/genet-cor.html">Genetic correlation across local ancestry segments</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/polygenic-scoring.html">Polygenic scoring for each local ancestry background</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/dataset.html">admix.Dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../notebooks/select-admixed.html">Selecting admixed individuals</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <h1>Source code for admix.simulate._geno</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span>
<span class="kn">import</span> <span class="nn">dask.array</span> <span class="k">as</span> <span class="nn">da</span>
<span class="kn">import</span> <span class="nn">admix</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">from</span> <span class="nn">._lanc</span> <span class="kn">import</span> <span class="n">hap_lanc</span> <span class="k">as</span> <span class="n">simulate_hap_lanc</span>


<div class="viewcode-block" id="admix_geno"><a class="viewcode-back" href="../../../_api/admix.simulate.admix_geno.html#admix.simulate.admix_geno">[docs]</a><span class="k">def</span> <span class="nf">admix_geno</span><span class="p">(</span>
    <span class="n">geno_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">da</span><span class="o">.</span><span class="n">Array</span><span class="p">],</span>
    <span class="n">df_snp</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">anc_props</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span>
    <span class="n">mosaic_size</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">n_indiv</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">return_sparse_lanc</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">admix</span><span class="o">.</span><span class="n">Dataset</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Simulate admixed genotype</span>

<span class="sd">    The generative model is:</span>

<span class="sd">    - for each ancestry, the allele frequencies are drawn</span>
<span class="sd">    - for each individual, breakpoints are drawn from a Poisson process. and the ancestry</span>
<span class="sd">        will be filled based a multinomial distribution with `n_anc` components</span>
<span class="sd">    - for each SNP, given the ancestry and the allele frequencies, the haplotype is drawn.</span>
<span class="sd">        Haplotype are simulated under some frequencies</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    geno_list : List[da.Array]</span>
<span class="sd">        List of ancestral data sets, each with (n_snp, n_indiv)</span>
<span class="sd">    df_snp : pd.DataFrame</span>
<span class="sd">        Dataframe of SNPs shared across ancestral data sets</span>
<span class="sd">    anc_props : list of float</span>
<span class="sd">        Proportion of ancestral populations</span>
<span class="sd">    mosaic_size : float</span>
<span class="sd">        Expected mosaic size in # of SNPs. use admix.lanc.calculate_mosaic_size() to</span>
<span class="sd">        calculate the mosaic size</span>
<span class="sd">    n_indiv : int</span>
<span class="sd">        Number of individuals to simulate</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    admix.Dataset</span>
<span class="sd">        Simulated admixed dataset</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">n_anc</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">geno_list</span><span class="p">)</span>
    <span class="n">n_snp</span> <span class="o">=</span> <span class="n">geno_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">assert</span> <span class="nb">all</span><span class="p">(</span>
        <span class="n">n_snp</span> <span class="o">==</span> <span class="n">geno</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">geno</span> <span class="ow">in</span> <span class="n">geno_list</span>
    <span class="p">),</span> <span class="s2">&quot;all geno must have the same number of SNPs&quot;</span>
    <span class="k">assert</span> <span class="n">n_snp</span> <span class="o">==</span> <span class="n">df_snp</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;df_snp must have the same number of SNPs&quot;</span>

    <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">anc_props</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;anc_props must sum to 1&quot;</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">anc_props</span><span class="p">)</span> <span class="o">==</span> <span class="n">n_anc</span><span class="p">,</span> <span class="s2">&quot;anc_props must have the same length as n_anc&quot;</span>
    <span class="n">anc_props</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">anc_props</span><span class="p">)</span>

    <span class="n">dset_hap_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">da</span><span class="o">.</span><span class="n">hstack</span><span class="p">([</span><span class="n">geno</span><span class="p">[:,</span> <span class="p">:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">geno</span><span class="p">[:,</span> <span class="p">:,</span> <span class="mi">1</span><span class="p">]])</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span> <span class="k">for</span> <span class="n">geno</span> <span class="ow">in</span> <span class="n">geno_list</span>
    <span class="p">]</span>

    <span class="n">hap_lanc_breaks</span><span class="p">,</span> <span class="n">hap_lanc_values</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">hap_lanc</span><span class="p">(</span>
        <span class="n">n_snp</span><span class="o">=</span><span class="n">n_snp</span><span class="p">,</span> <span class="n">n_hap</span><span class="o">=</span><span class="n">n_indiv</span> <span class="o">*</span> <span class="mi">2</span><span class="p">,</span> <span class="n">mosaic_size</span><span class="o">=</span><span class="n">mosaic_size</span><span class="p">,</span> <span class="n">anc_props</span><span class="o">=</span><span class="n">anc_props</span>
    <span class="p">)</span>
    <span class="n">geno</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_snp</span><span class="p">,</span> <span class="n">n_indiv</span> <span class="o">*</span> <span class="mi">2</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">hap_i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">n_indiv</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)):</span>
        <span class="n">start</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">stop</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">hap_lanc_breaks</span><span class="p">[</span><span class="n">hap_i</span><span class="p">],</span> <span class="n">hap_lanc_values</span><span class="p">[</span><span class="n">hap_i</span><span class="p">]):</span>
            <span class="n">geno</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="n">hap_i</span><span class="p">]</span> <span class="o">=</span> <span class="n">dset_hap_list</span><span class="p">[</span><span class="n">val</span><span class="p">][</span>
                <span class="n">start</span><span class="p">:</span><span class="n">stop</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">dset_hap_list</span><span class="p">[</span><span class="n">val</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="p">]</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">stop</span>
    <span class="n">lanc_breaks</span><span class="p">,</span> <span class="n">lanc_values</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">haplo2diplo</span><span class="p">(</span>
        <span class="n">breaks</span><span class="o">=</span><span class="n">hap_lanc_breaks</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="n">hap_lanc_values</span>
    <span class="p">)</span>
    <span class="n">lanc</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">Lanc</span><span class="p">(</span><span class="n">breaks</span><span class="o">=</span><span class="n">lanc_breaks</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="n">lanc_values</span><span class="p">)</span>

    <span class="c1"># a = np.random.randn(4, 5, 2)</span>
    <span class="c1"># b = np.zeros((4, 10))</span>
    <span class="c1"># b[:, 0::2] = a[:, :, 0]</span>
    <span class="c1"># b[:, 1::2] = a[:, :, 1]</span>
    <span class="c1"># assert np.all(b == a.reshape(4, 10))</span>
    <span class="c1"># assert np.all(b.reshape(4, 5, 2) == a)</span>

    <span class="c1"># equivalent: geno = np.dstack([geno[:, 0::2], geno[:, 1::2]])</span>
    <span class="n">geno</span> <span class="o">=</span> <span class="n">geno</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">n_snp</span><span class="p">,</span> <span class="n">n_indiv</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>

    <span class="n">dset</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span>
        <span class="n">geno</span><span class="o">=</span><span class="n">da</span><span class="o">.</span><span class="n">from_array</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span> <span class="n">chunks</span><span class="o">=-</span><span class="mi">1</span><span class="p">),</span>
        <span class="n">lanc</span><span class="o">=</span><span class="n">lanc</span><span class="o">.</span><span class="n">dask</span><span class="p">(</span><span class="n">snp_chunk</span><span class="o">=</span><span class="n">n_snp</span><span class="p">),</span>
        <span class="n">n_anc</span><span class="o">=</span><span class="n">n_anc</span><span class="p">,</span>
        <span class="n">snp</span><span class="o">=</span><span class="n">df_snp</span><span class="p">,</span>
        <span class="n">indiv</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">{</span><span class="s2">&quot;indiv&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;indiv_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">n_indiv</span><span class="p">)]}</span>
        <span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;indiv&quot;</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="n">return_sparse_lanc</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">dset</span><span class="p">,</span> <span class="n">lanc</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">dset</span></div>


<span class="k">def</span> <span class="nf">admix_geno_simple</span><span class="p">(</span>
    <span class="n">n_indiv</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">n_snp</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">n_anc</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">mosaic_size</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">anc_props</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">allele_freqs</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">af_maf_low</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
    <span class="n">af_maf_high</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">admix</span><span class="o">.</span><span class="n">Dataset</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Simulate admixed genotype</span>

<span class="sd">    The generative model is:</span>

<span class="sd">    - for each ancestry, the allele frequencies are drawn</span>
<span class="sd">    - for each individual, breakpoints are drawn from a Poisson process. and the ancestry will be filled based a multinomial distribution with `n_anc` components</span>
<span class="sd">    - for each SNP, given the ancestry and the allele frequencies, the haplotype is drawn. Haplotype are simulated under some frequencies</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    n_indiv : int</span>
<span class="sd">        Number of individuals</span>
<span class="sd">    n_snp : int</span>
<span class="sd">        Number of SNPs</span>
<span class="sd">    n_anc : int</span>
<span class="sd">        Number of ancestries</span>
<span class="sd">    mosaic_size : float</span>
<span class="sd">        Expected mosaic size in # of SNPs</span>
<span class="sd">    anc_props : list of float</span>
<span class="sd">        Proportion of ancestral populations, if not specified, the proportion</span>
<span class="sd">        is uniform over the ancestral populations.</span>
<span class="sd">    allele_freqs: List[np.ndarray]</span>
<span class="sd">        Allele frequencies, if not specified, the frequencies are drawn from</span>
<span class="sd">        using `af_maf_low` and `af_maf_high`</span>
<span class="sd">    af_maf_low : float</span>
<span class="sd">        Lowest allowed minor allele frequency</span>
<span class="sd">    af_maf_low : float</span>
<span class="sd">        Highest allowed minor allele frequency</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    admix.Dataset</span>
<span class="sd">        Simulated admixed dataset</span>


<span class="sd">    References</span>
<span class="sd">    ----------</span>
<span class="sd">    1. https://github.com/slowkoni/rfmix/blob/master/simulate.cpp</span>
<span class="sd">    2. https://github.com/slowkoni/admixture-simulation</span>
<span class="sd">    3. https://github.com/williamslab/admix-simu</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">anc_props</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">anc_props</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">n_anc</span><span class="p">)</span> <span class="o">/</span> <span class="n">n_anc</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">anc_props</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">anc_props</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">anc_props</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;anc_props must sum to 1&quot;</span>
    <span class="n">anc_props</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">anc_props</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">anc_props</span><span class="o">.</span><span class="n">size</span> <span class="o">==</span> <span class="n">n_anc</span><span class="p">,</span> <span class="s2">&quot;anc_props must have the same length as n_anc&quot;</span>

    <span class="n">breaks</span><span class="p">,</span> <span class="n">values</span> <span class="o">=</span> <span class="n">simulate_hap_lanc</span><span class="p">(</span>
        <span class="n">n_hap</span><span class="o">=</span><span class="n">n_indiv</span> <span class="o">*</span> <span class="mi">2</span><span class="p">,</span> <span class="n">n_snp</span><span class="o">=</span><span class="n">n_snp</span><span class="p">,</span> <span class="n">mosaic_size</span><span class="o">=</span><span class="n">mosaic_size</span><span class="p">,</span> <span class="n">anc_props</span><span class="o">=</span><span class="n">anc_props</span>
    <span class="p">)</span>
    <span class="n">breaks</span><span class="p">,</span> <span class="n">values</span> <span class="o">=</span> <span class="n">admix</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">haplo2diplo</span><span class="p">(</span><span class="n">breaks</span><span class="o">=</span><span class="n">breaks</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="n">values</span><span class="p">)</span>
    <span class="n">rls_lanc</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">admix</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">Lanc</span><span class="p">(</span><span class="n">breaks</span><span class="o">=</span><span class="n">breaks</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">dask</span><span class="p">()</span><span class="o">.</span><span class="n">swapaxes</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="c1"># n_indiv x n_snp x 2 (2 for each haplotype)</span>
    <span class="c1"># rls_lanc = np.dstack([rls_lanc[0:n_indiv, :], rls_lanc[n_indiv:, :]])</span>
    <span class="n">rls_geno</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">rls_lanc</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">allele_freqs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># allele frequencies for the two populations</span>
        <span class="n">allele_freqs</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="n">low</span><span class="o">=</span><span class="n">af_maf_low</span><span class="p">,</span> <span class="n">high</span><span class="o">=</span><span class="n">af_maf_high</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">n_snp</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_anc</span><span class="p">)</span>
        <span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">allele_freqs</span><span class="p">)</span> <span class="o">==</span> <span class="n">n_anc</span>
        <span class="p">),</span> <span class="s2">&quot;allele_freqs must have the same length as n_anc&quot;</span>
        <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span>
            <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">af</span><span class="p">)</span> <span class="o">==</span> <span class="n">n_snp</span> <span class="k">for</span> <span class="n">af</span> <span class="ow">in</span> <span class="n">allele_freqs</span><span class="p">]</span>
        <span class="p">),</span> <span class="s2">&quot;each element in allele_freqs must have the same length as n_snp&quot;</span>

    <span class="k">for</span> <span class="n">i_anc</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_anc</span><span class="p">):</span>
        <span class="n">rls_geno</span><span class="p">[</span><span class="n">rls_lanc</span> <span class="o">==</span> <span class="n">i_anc</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">binomial</span><span class="p">(</span>
            <span class="n">n</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="n">allele_freqs</span><span class="p">[</span><span class="n">i_anc</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">rls_lanc</span> <span class="o">==</span> <span class="n">i_anc</span><span class="p">)[</span><span class="mi">1</span><span class="p">]]</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">admix</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span>
        <span class="n">geno</span><span class="o">=</span><span class="n">da</span><span class="o">.</span><span class="n">from_array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">swapaxes</span><span class="p">(</span><span class="n">rls_geno</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">chunks</span><span class="o">=-</span><span class="mi">1</span><span class="p">),</span>
        <span class="n">lanc</span><span class="o">=</span><span class="n">da</span><span class="o">.</span><span class="n">from_array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">swapaxes</span><span class="p">(</span><span class="n">rls_lanc</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">chunks</span><span class="o">=-</span><span class="mi">1</span><span class="p">),</span>
        <span class="n">n_anc</span><span class="o">=</span><span class="n">n_anc</span><span class="p">,</span>
    <span class="p">)</span>


<span class="k">def</span> <span class="nf">admix_geno2</span><span class="p">(</span>
    <span class="n">n_indiv</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">n_snp</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">n_anc</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">mosaic_size</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">anc_props</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ancestral_af_range</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">],</span>
    <span class="n">F_st</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.3</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">admix</span><span class="o">.</span><span class="n">Dataset</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Simulate admixed genotype</span>
<span class="sd">    The generative model is:</span>
<span class="sd">        1. for each ancestry, the allele frequencies are drawn</span>
<span class="sd">        2. for each individual, breakpoints are drawn from a Poisson process.</span>
<span class="sd">            and the ancestry will be filled based a multinomial distribution with `n_anc`</span>
<span class="sd">            components</span>
<span class="sd">        3. for each SNP, given the ancestry and the allele frequencies, the haplotype</span>
<span class="sd">            is drawn</span>
<span class="sd">    Haplotype are simulated under some frequencies</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    n_indiv : int</span>
<span class="sd">        Number of individuals</span>
<span class="sd">    n_snp : int</span>
<span class="sd">        Number of SNPs</span>
<span class="sd">    n_anc : int</span>
<span class="sd">        Number of ancestries</span>
<span class="sd">    mosaic_size : float</span>
<span class="sd">        Expected mosaic size in # of SNPs</span>
<span class="sd">    anc_props : list of float</span>
<span class="sd">        Proportion of ancestral populations, if not specified, the proportion</span>
<span class="sd">        is uniform over the ancestral populations.</span>
<span class="sd">    ancestral_af_range: [float, float]</span>
<span class="sd">        allele frequencies range.</span>
<span class="sd">    F_st: float</span>
<span class="sd">        Distance for the two populations that form the admixed population.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">        admix.Dataset</span>
<span class="sd">        Simulated admixed dataset</span>

<span class="sd">    References</span>
<span class="sd">    ----------</span>
<span class="sd">    We follow Zaitlen et al. 2014, Nature Genetics</span>
<span class="sd">    See description in subsection &quot;Simulations with simulated genotypes&quot; in Methods</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">n_anc</span> <span class="o">==</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;n_anc must be 2, only two-way admixture is currently supported&quot;</span>
    <span class="c1"># global ancestry</span>
    <span class="k">if</span> <span class="n">anc_props</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">anc_props</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">n_anc</span><span class="p">)</span> <span class="o">/</span> <span class="n">n_anc</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">anc_props</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">anc_props</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">anc_props</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;anc_props must sum to 1&quot;</span>
    <span class="n">anc_props</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">anc_props</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">anc_props</span><span class="o">.</span><span class="n">size</span> <span class="o">==</span> <span class="n">n_anc</span><span class="p">,</span> <span class="s2">&quot;anc_props must have the same length as n_anc&quot;</span>

    <span class="c1"># local ancestry</span>
    <span class="n">rls_lanc</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">simulate_hap_lanc</span><span class="p">(</span>
            <span class="n">n_hap</span><span class="o">=</span><span class="n">n_indiv</span> <span class="o">*</span> <span class="mi">2</span><span class="p">,</span> <span class="n">n_snp</span><span class="o">=</span><span class="n">n_snp</span><span class="p">,</span> <span class="n">mosaic_size</span><span class="o">=</span><span class="n">mosaic_size</span><span class="p">,</span> <span class="n">anc_props</span><span class="o">=</span><span class="n">anc_props</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">dask</span><span class="p">()</span>
        <span class="o">.</span><span class="n">T</span>
    <span class="p">)</span>
    <span class="c1"># n_indiv x n_snp x 2 (2 for each haplotype)</span>
    <span class="n">rls_lanc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dstack</span><span class="p">([</span><span class="n">rls_lanc</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">n_indiv</span><span class="p">,</span> <span class="p">:],</span> <span class="n">rls_lanc</span><span class="p">[</span><span class="n">n_indiv</span><span class="p">:,</span> <span class="p">:]])</span>
    <span class="n">rls_geno</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">rls_lanc</span><span class="p">)</span>

    <span class="n">ancestral_afs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span>
        <span class="n">low</span><span class="o">=</span><span class="n">ancestral_af_range</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">high</span><span class="o">=</span><span class="n">ancestral_af_range</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="n">n_snp</span>
    <span class="p">)</span>

    <span class="c1"># allele frequencies for the two populations</span>
    <span class="n">allele_freqs</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">beta</span><span class="p">(</span>
            <span class="n">a</span><span class="o">=</span><span class="n">ancestral_afs</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">F_st</span><span class="p">)</span> <span class="o">/</span> <span class="n">F_st</span><span class="p">,</span>
            <span class="n">b</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">ancestral_afs</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">F_st</span><span class="p">)</span> <span class="o">/</span> <span class="n">F_st</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_anc</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="k">for</span> <span class="n">i_anc</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_anc</span><span class="p">):</span>
        <span class="n">rls_geno</span><span class="p">[</span><span class="n">rls_lanc</span> <span class="o">==</span> <span class="n">i_anc</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">binomial</span><span class="p">(</span>
            <span class="n">n</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="n">allele_freqs</span><span class="p">[</span><span class="n">i_anc</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">rls_lanc</span> <span class="o">==</span> <span class="n">i_anc</span><span class="p">)[</span><span class="mi">1</span><span class="p">]]</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">admix</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span>
        <span class="n">geno</span><span class="o">=</span><span class="n">da</span><span class="o">.</span><span class="n">from_array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">swapaxes</span><span class="p">(</span><span class="n">rls_geno</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)),</span>
        <span class="n">lanc</span><span class="o">=</span><span class="n">da</span><span class="o">.</span><span class="n">from_array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">swapaxes</span><span class="p">(</span><span class="n">rls_lanc</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)),</span>
        <span class="n">n_anc</span><span class="o">=</span><span class="n">n_anc</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2021-2024, Kangcheng Hou
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              <a class="muted-link " href="https://github.com/kangchenghou/admix-kit" aria-label="GitHub">
                <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 16 16">
                    <path fill-rule="evenodd" d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"></path>
                </svg>
            </a>
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js?v=b3ba4146"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="../../../_static/scripts/furo.js?v=32e29ea5"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/tabs.js?v=3ee01567"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>